{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "from repositories import retrieve_metadata\n",
    "from assess import assess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "keywords=[\"parkinson\"]\n",
    "repository_choice=\"5\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetching GDC metadata...\n",
      "Fetching metadata with keywords: ['suicide']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fetching metadata: 56809file [00:01, ?file/s]\n"
     ]
    },
    {
     "ename": "SSLError",
     "evalue": "HTTPSConnectionPool(host='api.gdc.cancer.gov', port=443): Max retries exceeded with url: /files?from=0&size=1000&sort=file_size%3Aasc&pretty=true (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self-signed certificate in certificate chain (_ssl.c:1007)')))",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mSSLCertVerificationError\u001b[0m                  Traceback (most recent call last)",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/urllib3/connectionpool.py:715\u001b[0m, in \u001b[0;36mHTTPConnectionPool.urlopen\u001b[0;34m(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, **response_kw)\u001b[0m\n\u001b[1;32m    714\u001b[0m \u001b[38;5;66;03m# Make the request on the httplib connection object.\u001b[39;00m\n\u001b[0;32m--> 715\u001b[0m httplib_response \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_make_request\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    716\u001b[0m \u001b[43m    \u001b[49m\u001b[43mconn\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    717\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    718\u001b[0m \u001b[43m    \u001b[49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    719\u001b[0m \u001b[43m    \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeout_obj\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    720\u001b[0m \u001b[43m    \u001b[49m\u001b[43mbody\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mbody\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    721\u001b[0m \u001b[43m    \u001b[49m\u001b[43mheaders\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mheaders\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    722\u001b[0m \u001b[43m    \u001b[49m\u001b[43mchunked\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mchunked\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    723\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    725\u001b[0m \u001b[38;5;66;03m# If we're going to release the connection in ``finally:``, then\u001b[39;00m\n\u001b[1;32m    726\u001b[0m \u001b[38;5;66;03m# the response doesn't need to know about the connection. Otherwise\u001b[39;00m\n\u001b[1;32m    727\u001b[0m \u001b[38;5;66;03m# it will also try to release it and we'll have a double-release\u001b[39;00m\n\u001b[1;32m    728\u001b[0m \u001b[38;5;66;03m# mess.\u001b[39;00m\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/urllib3/connectionpool.py:404\u001b[0m, in \u001b[0;36mHTTPConnectionPool._make_request\u001b[0;34m(self, conn, method, url, timeout, chunked, **httplib_request_kw)\u001b[0m\n\u001b[1;32m    403\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 404\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_validate_conn\u001b[49m\u001b[43m(\u001b[49m\u001b[43mconn\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    405\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m (SocketTimeout, BaseSSLError) \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m    406\u001b[0m     \u001b[38;5;66;03m# Py2 raises this as a BaseSSLError, Py3 raises it as socket timeout.\u001b[39;00m\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/urllib3/connectionpool.py:1058\u001b[0m, in \u001b[0;36mHTTPSConnectionPool._validate_conn\u001b[0;34m(self, conn)\u001b[0m\n\u001b[1;32m   1057\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28mgetattr\u001b[39m(conn, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msock\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m):  \u001b[38;5;66;03m# AppEngine might not have  `.sock`\u001b[39;00m\n\u001b[0;32m-> 1058\u001b[0m     \u001b[43mconn\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mconnect\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1060\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m conn\u001b[38;5;241m.\u001b[39mis_verified:\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/urllib3/connection.py:419\u001b[0m, in \u001b[0;36mHTTPSConnection.connect\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    417\u001b[0m     context\u001b[38;5;241m.\u001b[39mload_default_certs()\n\u001b[0;32m--> 419\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39msock \u001b[38;5;241m=\u001b[39m \u001b[43mssl_wrap_socket\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    420\u001b[0m \u001b[43m    \u001b[49m\u001b[43msock\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mconn\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    421\u001b[0m \u001b[43m    \u001b[49m\u001b[43mkeyfile\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mkey_file\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    422\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcertfile\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcert_file\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    423\u001b[0m \u001b[43m    \u001b[49m\u001b[43mkey_password\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mkey_password\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    424\u001b[0m \u001b[43m    \u001b[49m\u001b[43mca_certs\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mca_certs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    425\u001b[0m \u001b[43m    \u001b[49m\u001b[43mca_cert_dir\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mca_cert_dir\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    426\u001b[0m \u001b[43m    \u001b[49m\u001b[43mca_cert_data\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mca_cert_data\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    427\u001b[0m \u001b[43m    \u001b[49m\u001b[43mserver_hostname\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mserver_hostname\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    428\u001b[0m \u001b[43m    \u001b[49m\u001b[43mssl_context\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcontext\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    429\u001b[0m \u001b[43m    \u001b[49m\u001b[43mtls_in_tls\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtls_in_tls\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    430\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    432\u001b[0m \u001b[38;5;66;03m# If we're using all defaults and the connection\u001b[39;00m\n\u001b[1;32m    433\u001b[0m \u001b[38;5;66;03m# is TLSv1 or TLSv1.1 we throw a DeprecationWarning\u001b[39;00m\n\u001b[1;32m    434\u001b[0m \u001b[38;5;66;03m# for the host.\u001b[39;00m\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/urllib3/util/ssl_.py:449\u001b[0m, in \u001b[0;36mssl_wrap_socket\u001b[0;34m(sock, keyfile, certfile, cert_reqs, ca_certs, server_hostname, ssl_version, ciphers, ssl_context, ca_cert_dir, key_password, ca_cert_data, tls_in_tls)\u001b[0m\n\u001b[1;32m    448\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m send_sni:\n\u001b[0;32m--> 449\u001b[0m     ssl_sock \u001b[38;5;241m=\u001b[39m \u001b[43m_ssl_wrap_socket_impl\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    450\u001b[0m \u001b[43m        \u001b[49m\u001b[43msock\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcontext\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtls_in_tls\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mserver_hostname\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mserver_hostname\u001b[49m\n\u001b[1;32m    451\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    452\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/urllib3/util/ssl_.py:493\u001b[0m, in \u001b[0;36m_ssl_wrap_socket_impl\u001b[0;34m(sock, ssl_context, tls_in_tls, server_hostname)\u001b[0m\n\u001b[1;32m    492\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m server_hostname:\n\u001b[0;32m--> 493\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mssl_context\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mwrap_socket\u001b[49m\u001b[43m(\u001b[49m\u001b[43msock\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mserver_hostname\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mserver_hostname\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    494\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n",
      "File \u001b[0;32m/usr/lib/python3.10/ssl.py:513\u001b[0m, in \u001b[0;36mSSLContext.wrap_socket\u001b[0;34m(self, sock, server_side, do_handshake_on_connect, suppress_ragged_eofs, server_hostname, session)\u001b[0m\n\u001b[1;32m    507\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mwrap_socket\u001b[39m(\u001b[38;5;28mself\u001b[39m, sock, server_side\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m,\n\u001b[1;32m    508\u001b[0m                 do_handshake_on_connect\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m,\n\u001b[1;32m    509\u001b[0m                 suppress_ragged_eofs\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m,\n\u001b[1;32m    510\u001b[0m                 server_hostname\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, session\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m):\n\u001b[1;32m    511\u001b[0m     \u001b[38;5;66;03m# SSLSocket class handles server_hostname encoding before it calls\u001b[39;00m\n\u001b[1;32m    512\u001b[0m     \u001b[38;5;66;03m# ctx._wrap_socket()\u001b[39;00m\n\u001b[0;32m--> 513\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msslsocket_class\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_create\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    514\u001b[0m \u001b[43m        \u001b[49m\u001b[43msock\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msock\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    515\u001b[0m \u001b[43m        \u001b[49m\u001b[43mserver_side\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mserver_side\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    516\u001b[0m \u001b[43m        \u001b[49m\u001b[43mdo_handshake_on_connect\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdo_handshake_on_connect\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    517\u001b[0m \u001b[43m        \u001b[49m\u001b[43msuppress_ragged_eofs\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msuppress_ragged_eofs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    518\u001b[0m \u001b[43m        \u001b[49m\u001b[43mserver_hostname\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mserver_hostname\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    519\u001b[0m \u001b[43m        \u001b[49m\u001b[43mcontext\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    520\u001b[0m \u001b[43m        \u001b[49m\u001b[43msession\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msession\u001b[49m\n\u001b[1;32m    521\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/usr/lib/python3.10/ssl.py:1100\u001b[0m, in \u001b[0;36mSSLSocket._create\u001b[0;34m(cls, sock, server_side, do_handshake_on_connect, suppress_ragged_eofs, server_hostname, context, session)\u001b[0m\n\u001b[1;32m   1099\u001b[0m             \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdo_handshake_on_connect should not be specified for non-blocking sockets\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m-> 1100\u001b[0m         \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdo_handshake\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1101\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m (\u001b[38;5;167;01mOSError\u001b[39;00m, \u001b[38;5;167;01mValueError\u001b[39;00m):\n",
      "File \u001b[0;32m/usr/lib/python3.10/ssl.py:1371\u001b[0m, in \u001b[0;36mSSLSocket.do_handshake\u001b[0;34m(self, block)\u001b[0m\n\u001b[1;32m   1370\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39msettimeout(\u001b[38;5;28;01mNone\u001b[39;00m)\n\u001b[0;32m-> 1371\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_sslobj\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdo_handshake\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1372\u001b[0m \u001b[38;5;28;01mfinally\u001b[39;00m:\n",
      "\u001b[0;31mSSLCertVerificationError\u001b[0m: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self-signed certificate in certificate chain (_ssl.c:1007)",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mMaxRetryError\u001b[0m                             Traceback (most recent call last)",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/requests/adapters.py:440\u001b[0m, in \u001b[0;36mHTTPAdapter.send\u001b[0;34m(self, request, stream, timeout, verify, cert, proxies)\u001b[0m\n\u001b[1;32m    439\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m chunked:\n\u001b[0;32m--> 440\u001b[0m     resp \u001b[38;5;241m=\u001b[39m \u001b[43mconn\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43murlopen\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    441\u001b[0m \u001b[43m        \u001b[49m\u001b[43mmethod\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    442\u001b[0m \u001b[43m        \u001b[49m\u001b[43murl\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    443\u001b[0m \u001b[43m        \u001b[49m\u001b[43mbody\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mbody\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    444\u001b[0m \u001b[43m        \u001b[49m\u001b[43mheaders\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mheaders\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    445\u001b[0m \u001b[43m        \u001b[49m\u001b[43mredirect\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    446\u001b[0m \u001b[43m        \u001b[49m\u001b[43massert_same_host\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    447\u001b[0m \u001b[43m        \u001b[49m\u001b[43mpreload_content\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    448\u001b[0m \u001b[43m        \u001b[49m\u001b[43mdecode_content\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    449\u001b[0m \u001b[43m        \u001b[49m\u001b[43mretries\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmax_retries\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    450\u001b[0m \u001b[43m        \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeout\u001b[49m\n\u001b[1;32m    451\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    453\u001b[0m \u001b[38;5;66;03m# Send the request.\u001b[39;00m\n\u001b[1;32m    454\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/urllib3/connectionpool.py:799\u001b[0m, in \u001b[0;36mHTTPConnectionPool.urlopen\u001b[0;34m(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, **response_kw)\u001b[0m\n\u001b[1;32m    797\u001b[0m     e \u001b[38;5;241m=\u001b[39m ProtocolError(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mConnection aborted.\u001b[39m\u001b[38;5;124m\"\u001b[39m, e)\n\u001b[0;32m--> 799\u001b[0m retries \u001b[38;5;241m=\u001b[39m \u001b[43mretries\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mincrement\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    800\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43merror\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43me\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m_pool\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m_stacktrace\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msys\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mexc_info\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m2\u001b[39;49m\u001b[43m]\u001b[49m\n\u001b[1;32m    801\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    802\u001b[0m retries\u001b[38;5;241m.\u001b[39msleep()\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/urllib3/util/retry.py:592\u001b[0m, in \u001b[0;36mRetry.increment\u001b[0;34m(self, method, url, response, error, _pool, _stacktrace)\u001b[0m\n\u001b[1;32m    591\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m new_retry\u001b[38;5;241m.\u001b[39mis_exhausted():\n\u001b[0;32m--> 592\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m MaxRetryError(_pool, url, error \u001b[38;5;129;01mor\u001b[39;00m ResponseError(cause))\n\u001b[1;32m    594\u001b[0m log\u001b[38;5;241m.\u001b[39mdebug(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mIncremented Retry for (url=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m): \u001b[39m\u001b[38;5;132;01m%r\u001b[39;00m\u001b[38;5;124m\"\u001b[39m, url, new_retry)\n",
      "\u001b[0;31mMaxRetryError\u001b[0m: HTTPSConnectionPool(host='api.gdc.cancer.gov', port=443): Max retries exceeded with url: /files?from=0&size=1000&sort=file_size%3Aasc&pretty=true (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self-signed certificate in certificate chain (_ssl.c:1007)')))",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mSSLError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[11], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m metadata_list, request_status, url \u001b[38;5;241m=\u001b[39m \u001b[43mretrieve_metadata\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrepository_choice\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkeywords\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/Desktop/ft/FAIR/repositories.py:38\u001b[0m, in \u001b[0;36mretrieve_metadata\u001b[0;34m(repository_choice, keywords)\u001b[0m\n\u001b[1;32m     36\u001b[0m     metadata_list, request_status, url \u001b[38;5;241m=\u001b[39m fetch_experiments_metadata(keywords)\n\u001b[1;32m     37\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m repository_choice \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m5\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[0;32m---> 38\u001b[0m     metadata_list, request_status, url \u001b[38;5;241m=\u001b[39m \u001b[43mfetch_gdc_metadata\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkeywords\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     39\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m repository_choice \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m6\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[1;32m     40\u001b[0m     metadata_list, request_status, url \u001b[38;5;241m=\u001b[39m fetch_icgc_metadata(keywords)\n",
      "File \u001b[0;32m~/Desktop/ft/FAIR/repositories.py:338\u001b[0m, in \u001b[0;36mfetch_gdc_metadata\u001b[0;34m(keywords)\u001b[0m\n\u001b[1;32m    336\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mopen\u001b[39m(filename, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124ma\u001b[39m\u001b[38;5;124m'\u001b[39m) \u001b[38;5;28;01mas\u001b[39;00m file, tqdm(desc\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mFetching metadata\u001b[39m\u001b[38;5;124m\"\u001b[39m, unit\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mfile\u001b[39m\u001b[38;5;124m\"\u001b[39m, initial\u001b[38;5;241m=\u001b[39mtotal_files_fetched) \u001b[38;5;28;01mas\u001b[39;00m pbar:\n\u001b[1;32m    337\u001b[0m     \u001b[38;5;28;01mwhile\u001b[39;00m \u001b[38;5;28;01mTrue\u001b[39;00m:\n\u001b[0;32m--> 338\u001b[0m         response \u001b[38;5;241m=\u001b[39m \u001b[43mrequests\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[43mbase_url\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mparams\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mparams\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    339\u001b[0m         \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mAPI Call: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mresponse\u001b[38;5;241m.\u001b[39murl\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m    340\u001b[0m         request_status \u001b[38;5;241m=\u001b[39m response\u001b[38;5;241m.\u001b[39mstatus_code\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/requests/api.py:75\u001b[0m, in \u001b[0;36mget\u001b[0;34m(url, params, **kwargs)\u001b[0m\n\u001b[1;32m     64\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mget\u001b[39m(url, params\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[1;32m     65\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124mr\u001b[39m\u001b[38;5;124;03m\"\"\"Sends a GET request.\u001b[39;00m\n\u001b[1;32m     66\u001b[0m \n\u001b[1;32m     67\u001b[0m \u001b[38;5;124;03m    :param url: URL for the new :class:`Request` object.\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     72\u001b[0m \u001b[38;5;124;03m    :rtype: requests.Response\u001b[39;00m\n\u001b[1;32m     73\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m---> 75\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mrequest\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mget\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mparams\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mparams\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/requests/api.py:61\u001b[0m, in \u001b[0;36mrequest\u001b[0;34m(method, url, **kwargs)\u001b[0m\n\u001b[1;32m     57\u001b[0m \u001b[38;5;66;03m# By using the 'with' statement we are sure the session is closed, thus we\u001b[39;00m\n\u001b[1;32m     58\u001b[0m \u001b[38;5;66;03m# avoid leaving sockets open which can trigger a ResourceWarning in some\u001b[39;00m\n\u001b[1;32m     59\u001b[0m \u001b[38;5;66;03m# cases, and look like a memory leak in others.\u001b[39;00m\n\u001b[1;32m     60\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m sessions\u001b[38;5;241m.\u001b[39mSession() \u001b[38;5;28;01mas\u001b[39;00m session:\n\u001b[0;32m---> 61\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43msession\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmethod\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43murl\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/requests/sessions.py:529\u001b[0m, in \u001b[0;36mSession.request\u001b[0;34m(self, method, url, params, data, headers, cookies, files, auth, timeout, allow_redirects, proxies, hooks, stream, verify, cert, json)\u001b[0m\n\u001b[1;32m    524\u001b[0m send_kwargs \u001b[38;5;241m=\u001b[39m {\n\u001b[1;32m    525\u001b[0m     \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mtimeout\u001b[39m\u001b[38;5;124m'\u001b[39m: timeout,\n\u001b[1;32m    526\u001b[0m     \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mallow_redirects\u001b[39m\u001b[38;5;124m'\u001b[39m: allow_redirects,\n\u001b[1;32m    527\u001b[0m }\n\u001b[1;32m    528\u001b[0m send_kwargs\u001b[38;5;241m.\u001b[39mupdate(settings)\n\u001b[0;32m--> 529\u001b[0m resp \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msend\u001b[49m\u001b[43m(\u001b[49m\u001b[43mprep\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43msend_kwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    531\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m resp\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/requests/sessions.py:645\u001b[0m, in \u001b[0;36mSession.send\u001b[0;34m(self, request, **kwargs)\u001b[0m\n\u001b[1;32m    642\u001b[0m start \u001b[38;5;241m=\u001b[39m preferred_clock()\n\u001b[1;32m    644\u001b[0m \u001b[38;5;66;03m# Send the request\u001b[39;00m\n\u001b[0;32m--> 645\u001b[0m r \u001b[38;5;241m=\u001b[39m \u001b[43madapter\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msend\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrequest\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    647\u001b[0m \u001b[38;5;66;03m# Total elapsed time of the request (approximately)\u001b[39;00m\n\u001b[1;32m    648\u001b[0m elapsed \u001b[38;5;241m=\u001b[39m preferred_clock() \u001b[38;5;241m-\u001b[39m start\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/requests/adapters.py:517\u001b[0m, in \u001b[0;36mHTTPAdapter.send\u001b[0;34m(self, request, stream, timeout, verify, cert, proxies)\u001b[0m\n\u001b[1;32m    513\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m ProxyError(e, request\u001b[38;5;241m=\u001b[39mrequest)\n\u001b[1;32m    515\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(e\u001b[38;5;241m.\u001b[39mreason, _SSLError):\n\u001b[1;32m    516\u001b[0m         \u001b[38;5;66;03m# This branch is for urllib3 v1.22 and later.\u001b[39;00m\n\u001b[0;32m--> 517\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m SSLError(e, request\u001b[38;5;241m=\u001b[39mrequest)\n\u001b[1;32m    519\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mConnectionError\u001b[39;00m(e, request\u001b[38;5;241m=\u001b[39mrequest)\n\u001b[1;32m    521\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m ClosedPoolError \u001b[38;5;28;01mas\u001b[39;00m e:\n",
      "\u001b[0;31mSSLError\u001b[0m: HTTPSConnectionPool(host='api.gdc.cancer.gov', port=443): Max retries exceeded with url: /files?from=0&size=1000&sort=file_size%3Aasc&pretty=true (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self-signed certificate in certificate chain (_ssl.c:1007)')))"
     ]
    }
   ],
   "source": [
    "metadata_list, request_status, url = retrieve_metadata(repository_choice, keywords)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'initialSampleSize': '3,117 European ancestry Bipolar disorder cases, 1,273 European ancestry Major depressive disorder cases',\n",
       "  'gxe': False,\n",
       "  'gxg': False,\n",
       "  'snpCount': 1922309,\n",
       "  'qualifier': None,\n",
       "  'imputed': True,\n",
       "  'pooled': False,\n",
       "  'studyDesignComment': None,\n",
       "  'accessionId': 'GCST000854',\n",
       "  'fullPvalueSet': False,\n",
       "  'userRequested': False,\n",
       "  'platforms': [{'manufacturer': 'Affymetrix'}],\n",
       "  'ancestries': [{'type': 'initial',\n",
       "    'numberOfIndividuals': 4390,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'European'}],\n",
       "    'countryOfOrigin': [],\n",
       "    'countryOfRecruitment': [{'majorArea': 'Europe',\n",
       "      'region': 'Northern Europe',\n",
       "      'countryName': 'U.K.'}]},\n",
       "   {'type': 'replication',\n",
       "    'numberOfIndividuals': 2698,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'European'}],\n",
       "    'countryOfOrigin': [],\n",
       "    'countryOfRecruitment': [{'majorArea': 'Europe',\n",
       "      'region': 'Western Europe',\n",
       "      'countryName': 'Germany'}]},\n",
       "   {'type': 'replication',\n",
       "    'numberOfIndividuals': 1649,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'NR'}],\n",
       "    'countryOfOrigin': [],\n",
       "    'countryOfRecruitment': []}],\n",
       "  'diseaseTrait': {'trait': 'Suicide risk'},\n",
       "  'genotypingTechnologies': [{'genotypingTechnology': 'Genome-wide genotyping array'}],\n",
       "  'replicationSampleSize': '2,698 European ancestry Bipolar disorder cases, 1,649 Major depressive disorder cases',\n",
       "  'publicationInfo': {'pubmedId': '21041247',\n",
       "   'publicationDate': '2010-11-01',\n",
       "   'publication': 'Am J Psychiatry',\n",
       "   'title': 'Genome-wide association study of suicide attempts in mood disorder patients.',\n",
       "   'author': {'fullname': 'Perlis RH', 'orcid': None}},\n",
       "  '_links': {'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000854'},\n",
       "   'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000854{?projection}',\n",
       "    'templated': True},\n",
       "   'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000854/associations?projection=associationByStudy'},\n",
       "   'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000854/associations'},\n",
       "   'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000854/snps'},\n",
       "   'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000854/backgroundEfoTraits'},\n",
       "   'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000854/efoTraits'}}},\n",
       " {'initialSampleSize': '37 European ancestry cases, 36 European ancestry controls',\n",
       "  'gxe': False,\n",
       "  'gxg': False,\n",
       "  'snpCount': 251837,\n",
       "  'qualifier': None,\n",
       "  'imputed': False,\n",
       "  'pooled': False,\n",
       "  'studyDesignComment': None,\n",
       "  'accessionId': 'GCST000841',\n",
       "  'fullPvalueSet': False,\n",
       "  'userRequested': False,\n",
       "  'platforms': [{'manufacturer': 'Illumina'}],\n",
       "  'ancestries': [{'type': 'initial',\n",
       "    'numberOfIndividuals': 73,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'European'}],\n",
       "    'countryOfOrigin': [],\n",
       "    'countryOfRecruitment': [{'majorArea': 'Northern America',\n",
       "      'region': None,\n",
       "      'countryName': 'U.S.'}]}],\n",
       "  'diseaseTrait': {'trait': \"Dupuytren's disease\"},\n",
       "  'genotypingTechnologies': [{'genotypingTechnology': 'Genome-wide genotyping array'}],\n",
       "  'replicationSampleSize': 'NA',\n",
       "  'publicationInfo': {'pubmedId': '20971583',\n",
       "   'publicationDate': '2010-10-22',\n",
       "   'publication': 'J Hand Surg Am',\n",
       "   'title': \"Genome-wide association scan of Dupuytren's disease.\",\n",
       "   'author': {'fullname': 'Ojwang JO', 'orcid': None}},\n",
       "  '_links': {'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000841'},\n",
       "   'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000841{?projection}',\n",
       "    'templated': True},\n",
       "   'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000841/associations?projection=associationByStudy'},\n",
       "   'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000841/snps'},\n",
       "   'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000841/backgroundEfoTraits'},\n",
       "   'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000841/efoTraits'},\n",
       "   'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000841/associations'}}},\n",
       " {'initialSampleSize': 'up to 17,723 European ancestry individuals',\n",
       "  'gxe': False,\n",
       "  'gxg': False,\n",
       "  'snpCount': 2155369,\n",
       "  'qualifier': None,\n",
       "  'imputed': True,\n",
       "  'pooled': False,\n",
       "  'studyDesignComment': None,\n",
       "  'accessionId': 'GCST000809',\n",
       "  'fullPvalueSet': False,\n",
       "  'userRequested': False,\n",
       "  'platforms': [{'manufacturer': 'Affymetrix'},\n",
       "   {'manufacturer': 'Perlegen'},\n",
       "   {'manufacturer': 'Illumina'}],\n",
       "  'ancestries': [{'type': 'initial',\n",
       "    'numberOfIndividuals': 17723,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'European'}],\n",
       "    'countryOfOrigin': [],\n",
       "    'countryOfRecruitment': [{'majorArea': 'Europe',\n",
       "      'region': 'Northern Europe',\n",
       "      'countryName': 'Finland'},\n",
       "     {'majorArea': 'Europe',\n",
       "      'region': 'Southern Europe',\n",
       "      'countryName': 'Italy'},\n",
       "     {'majorArea': 'Europe',\n",
       "      'region': 'Northern Europe',\n",
       "      'countryName': 'U.K.'},\n",
       "     {'majorArea': 'Europe',\n",
       "      'region': 'Western Europe',\n",
       "      'countryName': 'Switzerland'}]},\n",
       "   {'type': 'replication',\n",
       "    'numberOfIndividuals': 37774,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'European'}],\n",
       "    'countryOfOrigin': [],\n",
       "    'countryOfRecruitment': [{'majorArea': 'Europe',\n",
       "      'region': 'Northern Europe',\n",
       "      'countryName': 'Finland'},\n",
       "     {'majorArea': 'Europe',\n",
       "      'region': 'Western Europe',\n",
       "      'countryName': 'Netherlands'},\n",
       "     {'majorArea': 'Europe',\n",
       "      'region': 'Northern Europe',\n",
       "      'countryName': 'U.K.'}]},\n",
       "   {'type': 'replication',\n",
       "    'numberOfIndividuals': 9665,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'South Asian'}],\n",
       "    'countryOfOrigin': [],\n",
       "    'countryOfRecruitment': [{'majorArea': 'Europe',\n",
       "      'region': 'Northern Europe',\n",
       "      'countryName': 'U.K.'}]}],\n",
       "  'diseaseTrait': {'trait': 'Triglycerides'},\n",
       "  'genotypingTechnologies': [{'genotypingTechnology': 'Genome-wide genotyping array'}],\n",
       "  'replicationSampleSize': 'up to 37,774 European ancestry individuals, up to 9,665 Indian Asian ancestry individuals',\n",
       "  'publicationInfo': {'pubmedId': '20864672',\n",
       "   'publicationDate': '2010-09-23',\n",
       "   'publication': 'Arterioscler Thromb Vasc Biol',\n",
       "   'title': 'Genetic variants influencing circulating lipid levels and risk of coronary artery disease.',\n",
       "   'author': {'fullname': 'Waterworth DM', 'orcid': None}},\n",
       "  '_links': {'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000809'},\n",
       "   'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000809{?projection}',\n",
       "    'templated': True},\n",
       "   'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000809/associations?projection=associationByStudy'},\n",
       "   'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000809/associations'},\n",
       "   'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000809/snps'},\n",
       "   'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000809/backgroundEfoTraits'},\n",
       "   'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000809/efoTraits'}}},\n",
       " {'initialSampleSize': '15,358 European ancestry individuals',\n",
       "  'gxe': False,\n",
       "  'gxg': False,\n",
       "  'snpCount': 2194468,\n",
       "  'qualifier': None,\n",
       "  'imputed': True,\n",
       "  'pooled': False,\n",
       "  'studyDesignComment': None,\n",
       "  'accessionId': 'GCST000847',\n",
       "  'fullPvalueSet': False,\n",
       "  'userRequested': False,\n",
       "  'platforms': [{'manufacturer': 'Affymetrix'}, {'manufacturer': 'Illumina'}],\n",
       "  'ancestries': [{'type': 'replication',\n",
       "    'numberOfIndividuals': 6652,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'European'}],\n",
       "    'countryOfOrigin': [],\n",
       "    'countryOfRecruitment': [{'majorArea': 'Northern America',\n",
       "      'region': None,\n",
       "      'countryName': 'U.S.'},\n",
       "     {'majorArea': 'Oceania',\n",
       "      'region': 'Australia/New Zealand',\n",
       "      'countryName': 'Australia'}]},\n",
       "   {'type': 'initial',\n",
       "    'numberOfIndividuals': 15358,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'European'}],\n",
       "    'countryOfOrigin': [],\n",
       "    'countryOfRecruitment': [{'majorArea': 'Northern America',\n",
       "      'region': None,\n",
       "      'countryName': 'U.S.'},\n",
       "     {'majorArea': 'Europe',\n",
       "      'region': 'Western Europe',\n",
       "      'countryName': 'Netherlands'}]}],\n",
       "  'diseaseTrait': {'trait': 'Retinal vascular caliber'},\n",
       "  'genotypingTechnologies': [{'genotypingTechnology': 'Genome-wide genotyping array'}],\n",
       "  'replicationSampleSize': '6,652 European ancestry individuals',\n",
       "  'publicationInfo': {'pubmedId': '21060863',\n",
       "   'publicationDate': '2010-10-28',\n",
       "   'publication': 'PLoS Genet',\n",
       "   'title': 'Four novel Loci (19q13, 6q24, 12q24, and 5q14) influence the microcirculation in vivo.',\n",
       "   'author': {'fullname': 'Ikram MK', 'orcid': None}},\n",
       "  '_links': {'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000847'},\n",
       "   'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000847{?projection}',\n",
       "    'templated': True},\n",
       "   'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000847/associations?projection=associationByStudy'},\n",
       "   'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000847/snps'},\n",
       "   'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000847/backgroundEfoTraits'},\n",
       "   'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000847/efoTraits'},\n",
       "   'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000847/associations'}}},\n",
       " {'initialSampleSize': '105 East Asian ancestry cases',\n",
       "  'gxe': False,\n",
       "  'gxg': False,\n",
       "  'snpCount': 109365,\n",
       "  'qualifier': None,\n",
       "  'imputed': False,\n",
       "  'pooled': False,\n",
       "  'studyDesignComment': None,\n",
       "  'accessionId': 'GCST000871',\n",
       "  'fullPvalueSet': False,\n",
       "  'userRequested': False,\n",
       "  'platforms': [{'manufacturer': 'Illumina'}],\n",
       "  'ancestries': [{'type': 'initial',\n",
       "    'numberOfIndividuals': 105,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'East Asian'}],\n",
       "    'countryOfOrigin': [],\n",
       "    'countryOfRecruitment': [{'majorArea': 'Asia',\n",
       "      'region': 'Eastern Asia',\n",
       "      'countryName': 'Japan'}]}],\n",
       "  'diseaseTrait': {'trait': 'Non-small cell lung cancer (survival)'},\n",
       "  'genotypingTechnologies': [{'genotypingTechnology': 'Genome-wide genotyping array'}],\n",
       "  'replicationSampleSize': 'NA',\n",
       "  'publicationInfo': {'pubmedId': '21079520',\n",
       "   'publicationDate': '2010-11-12',\n",
       "   'publication': 'J Thorac Oncol',\n",
       "   'title': 'Genome-wide association study on overall survival of advanced non-small cell lung cancer patients treated with carboplatin and paclitaxel.',\n",
       "   'author': {'fullname': 'Sato Y', 'orcid': None}},\n",
       "  '_links': {'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000871'},\n",
       "   'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000871{?projection}',\n",
       "    'templated': True},\n",
       "   'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000871/associations?projection=associationByStudy'},\n",
       "   'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000871/snps'},\n",
       "   'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000871/backgroundEfoTraits'},\n",
       "   'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000871/efoTraits'},\n",
       "   'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000871/associations'}}},\n",
       " {'initialSampleSize': '12,441 European ancestry children',\n",
       "  'gxe': False,\n",
       "  'gxg': False,\n",
       "  'snpCount': 138093,\n",
       "  'qualifier': None,\n",
       "  'imputed': True,\n",
       "  'pooled': False,\n",
       "  'studyDesignComment': None,\n",
       "  'accessionId': 'GCST001837',\n",
       "  'fullPvalueSet': True,\n",
       "  'userRequested': False,\n",
       "  'platforms': [{'manufacturer': 'Affymetrix'}, {'manufacturer': 'Illumina'}],\n",
       "  'ancestries': [{'type': 'replication',\n",
       "    'numberOfIndividuals': 5548,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'European'}],\n",
       "    'countryOfOrigin': [],\n",
       "    'countryOfRecruitment': [{'majorArea': 'Northern America',\n",
       "      'region': None,\n",
       "      'countryName': 'U.S.'},\n",
       "     {'majorArea': 'Europe',\n",
       "      'region': 'Western Europe',\n",
       "      'countryName': 'Netherlands'}]},\n",
       "   {'type': 'initial',\n",
       "    'numberOfIndividuals': 12441,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'European'}],\n",
       "    'countryOfOrigin': [],\n",
       "    'countryOfRecruitment': [{'majorArea': 'Oceania',\n",
       "      'region': 'Australia/New Zealand',\n",
       "      'countryName': 'Australia'},\n",
       "     {'majorArea': 'Europe',\n",
       "      'region': 'Northern Europe',\n",
       "      'countryName': 'U.K.'}]}],\n",
       "  'diseaseTrait': {'trait': 'Intelligence (childhood)'},\n",
       "  'genotypingTechnologies': [{'genotypingTechnology': 'Genome-wide genotyping array'}],\n",
       "  'replicationSampleSize': '5,548 European ancestry children',\n",
       "  'publicationInfo': {'pubmedId': '23358156',\n",
       "   'publicationDate': '2013-01-29',\n",
       "   'publication': 'Mol Psychiatry',\n",
       "   'title': 'Childhood intelligence is heritable, highly polygenic and associated with FNBP1L.',\n",
       "   'author': {'fullname': 'Benyamin B', 'orcid': '0000-0001-5608-2293'}},\n",
       "  '_links': {'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001837'},\n",
       "   'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001837{?projection}',\n",
       "    'templated': True},\n",
       "   'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001837/associations?projection=associationByStudy'},\n",
       "   'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001837/associations'},\n",
       "   'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001837/snps'},\n",
       "   'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001837/backgroundEfoTraits'},\n",
       "   'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001837/efoTraits'}}},\n",
       " {'initialSampleSize': '2,132 Indian ancestry individuals, 2,313 Malay ancestry individuals',\n",
       "  'gxe': False,\n",
       "  'gxg': False,\n",
       "  'snpCount': 551808,\n",
       "  'qualifier': None,\n",
       "  'imputed': False,\n",
       "  'pooled': False,\n",
       "  'studyDesignComment': None,\n",
       "  'accessionId': 'GCST000970',\n",
       "  'fullPvalueSet': False,\n",
       "  'userRequested': False,\n",
       "  'platforms': [{'manufacturer': 'Illumina'}],\n",
       "  'ancestries': [{'type': 'initial',\n",
       "    'numberOfIndividuals': 2313,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'South East Asian'}],\n",
       "    'countryOfOrigin': [{'majorArea': 'NR',\n",
       "      'region': 'NR',\n",
       "      'countryName': 'NR'}],\n",
       "    'countryOfRecruitment': [{'majorArea': 'Asia',\n",
       "      'region': 'South-eastern Asia',\n",
       "      'countryName': 'Singapore'}]},\n",
       "   {'type': 'initial',\n",
       "    'numberOfIndividuals': 2132,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'South Asian'}],\n",
       "    'countryOfOrigin': [{'majorArea': 'NR',\n",
       "      'region': 'NR',\n",
       "      'countryName': 'NR'}],\n",
       "    'countryOfRecruitment': [{'majorArea': 'Asia',\n",
       "      'region': 'South-eastern Asia',\n",
       "      'countryName': 'Singapore'}]},\n",
       "   {'type': 'replication',\n",
       "    'numberOfIndividuals': 9326,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'European'}],\n",
       "    'countryOfOrigin': [{'majorArea': 'NR',\n",
       "      'region': 'NR',\n",
       "      'countryName': 'NR'}],\n",
       "    'countryOfRecruitment': [{'majorArea': 'Europe',\n",
       "      'region': 'Western Europe',\n",
       "      'countryName': 'Netherlands'}]}],\n",
       "  'diseaseTrait': {'trait': 'Optic disc area'},\n",
       "  'genotypingTechnologies': [{'genotypingTechnology': 'Genome-wide genotyping array'}],\n",
       "  'replicationSampleSize': '9,326 European ancestry individuals',\n",
       "  'publicationInfo': {'pubmedId': '21307088',\n",
       "   'publicationDate': '2011-02-09',\n",
       "   'publication': 'Hum Mol Genet',\n",
       "   'title': 'Genome-wide association studies in Asians confirm the involvement of ATOH7 and TGFBR3, and further identify CARD10 as a novel locus influencing optic disc area.',\n",
       "   'author': {'fullname': 'Khor CC', 'orcid': '0000-0002-1128-4729'}},\n",
       "  '_links': {'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000970'},\n",
       "   'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000970{?projection}',\n",
       "    'templated': True},\n",
       "   'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000970/associations?projection=associationByStudy'},\n",
       "   'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000970/snps'},\n",
       "   'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000970/backgroundEfoTraits'},\n",
       "   'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000970/efoTraits'},\n",
       "   'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000970/associations'}}},\n",
       " {'initialSampleSize': '1,339 European ancestry cases',\n",
       "  'gxe': False,\n",
       "  'gxg': False,\n",
       "  'snpCount': 520460,\n",
       "  'qualifier': None,\n",
       "  'imputed': False,\n",
       "  'pooled': False,\n",
       "  'studyDesignComment': None,\n",
       "  'accessionId': 'GCST000981',\n",
       "  'fullPvalueSet': False,\n",
       "  'userRequested': False,\n",
       "  'platforms': [{'manufacturer': 'Illumina'}],\n",
       "  'ancestries': [{'type': 'initial',\n",
       "    'numberOfIndividuals': 1339,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'European'}],\n",
       "    'countryOfOrigin': [],\n",
       "    'countryOfRecruitment': []},\n",
       "   {'type': 'replication',\n",
       "    'numberOfIndividuals': 677,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'European'}],\n",
       "    'countryOfOrigin': [],\n",
       "    'countryOfRecruitment': []}],\n",
       "  'diseaseTrait': {'trait': 'Vitiligo'},\n",
       "  'genotypingTechnologies': [{'genotypingTechnology': 'Genome-wide genotyping array'}],\n",
       "  'replicationSampleSize': '677 European ancestry cases',\n",
       "  'publicationInfo': {'pubmedId': '21326295',\n",
       "   'publicationDate': '2011-02-17',\n",
       "   'publication': 'J Invest Dermatol',\n",
       "   'title': 'Genome-wide analysis identifies a quantitative trait locus in the MHC class II region associated with generalized vitiligo age of onset.',\n",
       "   'author': {'fullname': 'Jin Y', 'orcid': None}},\n",
       "  '_links': {'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000981'},\n",
       "   'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000981{?projection}',\n",
       "    'templated': True},\n",
       "   'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000981/associations?projection=associationByStudy'},\n",
       "   'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000981/associations'},\n",
       "   'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000981/snps'},\n",
       "   'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000981/backgroundEfoTraits'},\n",
       "   'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000981/efoTraits'}}},\n",
       " {'initialSampleSize': '31,580 European ancestry individuals',\n",
       "  'gxe': False,\n",
       "  'gxg': False,\n",
       "  'snpCount': 2500000,\n",
       "  'qualifier': '~',\n",
       "  'imputed': True,\n",
       "  'pooled': False,\n",
       "  'studyDesignComment': None,\n",
       "  'accessionId': 'GCST000988',\n",
       "  'fullPvalueSet': True,\n",
       "  'userRequested': False,\n",
       "  'platforms': [{'manufacturer': 'Illumina'}, {'manufacturer': 'Affymetrix'}],\n",
       "  'ancestries': [{'type': 'initial',\n",
       "    'numberOfIndividuals': 31580,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'European'}],\n",
       "    'countryOfOrigin': [],\n",
       "    'countryOfRecruitment': [{'majorArea': 'Northern America',\n",
       "      'region': None,\n",
       "      'countryName': 'U.S.'},\n",
       "     {'majorArea': 'Europe',\n",
       "      'region': 'Southern Europe',\n",
       "      'countryName': 'Italy'},\n",
       "     {'majorArea': 'Europe',\n",
       "      'region': 'Western Europe',\n",
       "      'countryName': 'Germany'},\n",
       "     {'majorArea': 'Europe',\n",
       "      'region': 'Northern Europe',\n",
       "      'countryName': 'U.K.'},\n",
       "     {'majorArea': 'Europe',\n",
       "      'region': 'Western Europe',\n",
       "      'countryName': 'Switzerland'}]},\n",
       "   {'type': 'replication',\n",
       "    'numberOfIndividuals': 31277,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'European'}],\n",
       "    'countryOfOrigin': [],\n",
       "    'countryOfRecruitment': [{'majorArea': 'Northern America',\n",
       "      'region': None,\n",
       "      'countryName': 'U.S.'},\n",
       "     {'majorArea': 'Europe',\n",
       "      'region': 'Northern Europe',\n",
       "      'countryName': 'Iceland'},\n",
       "     {'majorArea': 'Europe',\n",
       "      'region': 'Western Europe',\n",
       "      'countryName': 'Germany'},\n",
       "     {'majorArea': 'Europe',\n",
       "      'region': 'Southern Europe',\n",
       "      'countryName': 'Croatia'}]}],\n",
       "  'diseaseTrait': {'trait': 'Urinary albumin excretion'},\n",
       "  'genotypingTechnologies': [{'genotypingTechnology': 'Genome-wide genotyping array'}],\n",
       "  'replicationSampleSize': '31,277 European ancestry individuals',\n",
       "  'publicationInfo': {'pubmedId': '21355061',\n",
       "   'publicationDate': '2011-02-25',\n",
       "   'publication': 'J Am Soc Nephrol',\n",
       "   'title': 'CUBN is a gene locus for albuminuria.',\n",
       "   'author': {'fullname': 'Bger CA', 'orcid': None}},\n",
       "  '_links': {'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000988'},\n",
       "   'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000988{?projection}',\n",
       "    'templated': True},\n",
       "   'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000988/associations?projection=associationByStudy'},\n",
       "   'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000988/snps'},\n",
       "   'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000988/backgroundEfoTraits'},\n",
       "   'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000988/efoTraits'},\n",
       "   'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000988/associations'}}},\n",
       " {'initialSampleSize': '7,473 African American individuals',\n",
       "  'gxe': False,\n",
       "  'gxg': False,\n",
       "  'snpCount': 2500000,\n",
       "  'qualifier': None,\n",
       "  'imputed': True,\n",
       "  'pooled': False,\n",
       "  'studyDesignComment': None,\n",
       "  'accessionId': 'GCST000997',\n",
       "  'fullPvalueSet': False,\n",
       "  'userRequested': False,\n",
       "  'platforms': [{'manufacturer': 'Affymetrix'}],\n",
       "  'ancestries': [{'type': 'initial',\n",
       "    'numberOfIndividuals': 7473,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'African American or Afro-Caribbean'}],\n",
       "    'countryOfOrigin': [],\n",
       "    'countryOfRecruitment': [{'majorArea': 'Northern America',\n",
       "      'region': None,\n",
       "      'countryName': 'U.S.'}]},\n",
       "   {'type': 'replication',\n",
       "    'numberOfIndividuals': 10694,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'African American or Afro-Caribbean'}],\n",
       "    'countryOfOrigin': [{'majorArea': 'NR',\n",
       "      'region': 'NR',\n",
       "      'countryName': 'NR'}],\n",
       "    'countryOfRecruitment': [{'majorArea': 'Northern America',\n",
       "      'region': None,\n",
       "      'countryName': 'U.S.'}]},\n",
       "   {'type': 'replication',\n",
       "    'numberOfIndividuals': 69899,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'European'}],\n",
       "    'countryOfOrigin': [],\n",
       "    'countryOfRecruitment': [{'majorArea': 'Northern America',\n",
       "      'region': None,\n",
       "      'countryName': 'U.S.'}]},\n",
       "   {'type': 'replication',\n",
       "    'numberOfIndividuals': 1188,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'Sub-Saharan African'}],\n",
       "    'countryOfOrigin': [{'majorArea': 'NR',\n",
       "      'region': 'NR',\n",
       "      'countryName': 'NR'}],\n",
       "    'countryOfRecruitment': [{'majorArea': 'Africa',\n",
       "      'region': 'Western Africa',\n",
       "      'countryName': 'Nigeria'},\n",
       "     {'majorArea': 'Africa',\n",
       "      'region': 'Western Africa',\n",
       "      'countryName': 'Niger'}]}],\n",
       "  'diseaseTrait': {'trait': 'Blood pressure'},\n",
       "  'genotypingTechnologies': [{'genotypingTechnology': 'Genome-wide genotyping array'}],\n",
       "  'replicationSampleSize': '1,188 Sub-Saharan African individuals, 10,694 African American individuals, 69,899 European ancestry individuals',\n",
       "  'publicationInfo': {'pubmedId': '21378095',\n",
       "   'publicationDate': '2011-03-04',\n",
       "   'publication': 'Hum Mol Genet',\n",
       "   'title': 'Association of genetic variation with systolic and diastolic blood pressure among African Americans: the Candidate Gene Association Resource study.',\n",
       "   'author': {'fullname': 'Fox ER', 'orcid': None}},\n",
       "  '_links': {'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000997'},\n",
       "   'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000997{?projection}',\n",
       "    'templated': True},\n",
       "   'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000997/associations?projection=associationByStudy'},\n",
       "   'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000997/associations'},\n",
       "   'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000997/snps'},\n",
       "   'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000997/backgroundEfoTraits'},\n",
       "   'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000997/efoTraits'}}},\n",
       " {'initialSampleSize': '811 anti-dsDNA positive European ancestry cases, 906 anti-dsDNA negative European ancestry cases, 4,813 European ancestry controls',\n",
       "  'gxe': False,\n",
       "  'gxg': False,\n",
       "  'snpCount': 421318,\n",
       "  'qualifier': None,\n",
       "  'imputed': True,\n",
       "  'pooled': False,\n",
       "  'studyDesignComment': None,\n",
       "  'accessionId': 'GCST000996',\n",
       "  'fullPvalueSet': False,\n",
       "  'userRequested': False,\n",
       "  'platforms': [{'manufacturer': 'Illumina'}],\n",
       "  'ancestries': [{'type': 'initial',\n",
       "    'numberOfIndividuals': 6530,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'European'}],\n",
       "    'countryOfOrigin': [],\n",
       "    'countryOfRecruitment': [{'majorArea': 'Northern America',\n",
       "      'region': None,\n",
       "      'countryName': 'U.S.'}]}],\n",
       "  'diseaseTrait': {'trait': 'Systemic lupus erythematosus'},\n",
       "  'genotypingTechnologies': [{'genotypingTechnology': 'Genome-wide genotyping array'}],\n",
       "  'replicationSampleSize': 'NA',\n",
       "  'publicationInfo': {'pubmedId': '21408207',\n",
       "   'publicationDate': '2011-03-03',\n",
       "   'publication': 'PLoS Genet',\n",
       "   'title': 'Differential genetic associations for systemic lupus erythematosus based on anti-dsDNA autoantibody production.',\n",
       "   'author': {'fullname': 'Chung SA', 'orcid': None}},\n",
       "  '_links': {'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000996'},\n",
       "   'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000996{?projection}',\n",
       "    'templated': True},\n",
       "   'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000996/associations?projection=associationByStudy'},\n",
       "   'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000996/snps'},\n",
       "   'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000996/backgroundEfoTraits'},\n",
       "   'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000996/efoTraits'},\n",
       "   'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000996/associations'}}},\n",
       " {'initialSampleSize': '1,831 European ancestry cases, 1,764 European ancestry controls',\n",
       "  'gxe': False,\n",
       "  'gxg': False,\n",
       "  'snpCount': None,\n",
       "  'qualifier': None,\n",
       "  'imputed': True,\n",
       "  'pooled': False,\n",
       "  'studyDesignComment': None,\n",
       "  'accessionId': 'GCST000986',\n",
       "  'fullPvalueSet': False,\n",
       "  'userRequested': False,\n",
       "  'platforms': [{'manufacturer': 'Affymetrix'}, {'manufacturer': 'Illumina'}],\n",
       "  'ancestries': [{'type': 'initial',\n",
       "    'numberOfIndividuals': 3595,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'European'}],\n",
       "    'countryOfOrigin': [],\n",
       "    'countryOfRecruitment': []},\n",
       "   {'type': 'replication',\n",
       "    'numberOfIndividuals': 1502,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'NR'}],\n",
       "    'countryOfOrigin': [],\n",
       "    'countryOfRecruitment': []}],\n",
       "  'diseaseTrait': {'trait': \"Alzheimer's disease (late onset)\"},\n",
       "  'genotypingTechnologies': [{'genotypingTechnology': 'Genome-wide genotyping array'}],\n",
       "  'replicationSampleSize': '751 cases, 751 controls',\n",
       "  'publicationInfo': {'pubmedId': '21390209',\n",
       "   'publicationDate': '2011-02-24',\n",
       "   'publication': 'PLoS One',\n",
       "   'title': \"Meta-analysis for genome-wide association study identifies multiple variants at the BIN1 locus associated with late-onset Alzheimer's disease.\",\n",
       "   'author': {'fullname': 'Hu X', 'orcid': None}},\n",
       "  '_links': {'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000986'},\n",
       "   'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000986{?projection}',\n",
       "    'templated': True},\n",
       "   'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000986/associations?projection=associationByStudy'},\n",
       "   'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000986/associations'},\n",
       "   'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000986/snps'},\n",
       "   'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000986/backgroundEfoTraits'},\n",
       "   'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000986/efoTraits'}}},\n",
       " {'initialSampleSize': '880 Amish individuals, 6,296 European ancestry individuals',\n",
       "  'gxe': False,\n",
       "  'gxg': False,\n",
       "  'snpCount': 2400000,\n",
       "  'qualifier': '~',\n",
       "  'imputed': True,\n",
       "  'pooled': False,\n",
       "  'studyDesignComment': None,\n",
       "  'accessionId': 'GCST001008',\n",
       "  'fullPvalueSet': False,\n",
       "  'userRequested': False,\n",
       "  'platforms': [{'manufacturer': 'Affymetrix'}, {'manufacturer': 'Illumina'}],\n",
       "  'ancestries': [{'type': 'initial',\n",
       "    'numberOfIndividuals': 7176,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'European'}],\n",
       "    'countryOfOrigin': [],\n",
       "    'countryOfRecruitment': [{'majorArea': 'Northern America',\n",
       "      'region': None,\n",
       "      'countryName': 'U.S.'},\n",
       "     {'majorArea': 'Europe',\n",
       "      'region': 'Northern Europe',\n",
       "      'countryName': 'Iceland'}]},\n",
       "   {'type': 'replication',\n",
       "    'numberOfIndividuals': 1997,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'European'}],\n",
       "    'countryOfOrigin': [],\n",
       "    'countryOfRecruitment': [{'majorArea': 'Northern America',\n",
       "      'region': None,\n",
       "      'countryName': 'U.S.'}]}],\n",
       "  'diseaseTrait': {'trait': 'Nonalcoholic fatty liver disease'},\n",
       "  'genotypingTechnologies': [{'genotypingTechnology': 'Genome-wide genotyping array'}],\n",
       "  'replicationSampleSize': '592 European ancestry cases, 1,405 European ancestry controls',\n",
       "  'publicationInfo': {'pubmedId': '21423719',\n",
       "   'publicationDate': '2011-03-10',\n",
       "   'publication': 'PLoS Genet',\n",
       "   'title': 'Genome-wide association analysis identifies variants associated with nonalcoholic fatty liver disease that have distinct effects on metabolic traits.',\n",
       "   'author': {'fullname': 'Speliotes EK', 'orcid': None}},\n",
       "  '_links': {'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001008'},\n",
       "   'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001008{?projection}',\n",
       "    'templated': True},\n",
       "   'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001008/associations?projection=associationByStudy'},\n",
       "   'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001008/snps'},\n",
       "   'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001008/backgroundEfoTraits'},\n",
       "   'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001008/efoTraits'},\n",
       "   'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001008/associations'}}},\n",
       " {'initialSampleSize': '229 Hispanic individuals from 34 families',\n",
       "  'gxe': False,\n",
       "  'gxg': False,\n",
       "  'snpCount': 309200,\n",
       "  'qualifier': None,\n",
       "  'imputed': False,\n",
       "  'pooled': False,\n",
       "  'studyDesignComment': None,\n",
       "  'accessionId': 'GCST000711',\n",
       "  'fullPvalueSet': False,\n",
       "  'userRequested': False,\n",
       "  'platforms': [{'manufacturer': 'Illumina'}],\n",
       "  'ancestries': [{'type': 'initial',\n",
       "    'numberOfIndividuals': 229,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'Hispanic or Latin American'}],\n",
       "    'countryOfOrigin': [],\n",
       "    'countryOfRecruitment': [{'majorArea': 'Northern America',\n",
       "      'region': None,\n",
       "      'countryName': 'U.S.'}]},\n",
       "   {'type': 'replication',\n",
       "    'numberOfIndividuals': 961,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'Hispanic or Latin American'}],\n",
       "    'countryOfOrigin': [],\n",
       "    'countryOfRecruitment': [{'majorArea': 'Northern America',\n",
       "      'region': None,\n",
       "      'countryName': 'U.S.'}]}],\n",
       "  'diseaseTrait': {'trait': 'Vitamin D levels'},\n",
       "  'genotypingTechnologies': [{'genotypingTechnology': 'Genome-wide genotyping array'}],\n",
       "  'replicationSampleSize': '961 Hispanic individuals',\n",
       "  'publicationInfo': {'pubmedId': '20600896',\n",
       "   'publicationDate': '2010-06-26',\n",
       "   'publication': 'J Steroid Biochem Mol Biol',\n",
       "   'title': 'Genome-wide association study of vitamin D concentrations in Hispanic Americans: the IRAS family study.',\n",
       "   'author': {'fullname': 'Engelman CD', 'orcid': None}},\n",
       "  '_links': {'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000711'},\n",
       "   'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000711{?projection}',\n",
       "    'templated': True},\n",
       "   'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000711/associations?projection=associationByStudy'},\n",
       "   'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000711/associations'},\n",
       "   'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000711/snps'},\n",
       "   'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000711/backgroundEfoTraits'},\n",
       "   'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000711/efoTraits'}}},\n",
       " {'initialSampleSize': '51 European ancestry rapid progressor male cases, 57 European ancestry moderate progressor male cases, 48 European ancestry long-term progressor male cases',\n",
       "  'gxe': False,\n",
       "  'gxg': False,\n",
       "  'snpCount': 345926,\n",
       "  'qualifier': None,\n",
       "  'imputed': False,\n",
       "  'pooled': False,\n",
       "  'studyDesignComment': None,\n",
       "  'accessionId': 'GCST000596',\n",
       "  'fullPvalueSet': False,\n",
       "  'userRequested': False,\n",
       "  'platforms': [{'manufacturer': 'Affymetrix'}],\n",
       "  'ancestries': [{'type': 'initial',\n",
       "    'numberOfIndividuals': 156,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'European'}],\n",
       "    'countryOfOrigin': [],\n",
       "    'countryOfRecruitment': [{'majorArea': 'Northern America',\n",
       "      'region': None,\n",
       "      'countryName': 'U.S.'}]},\n",
       "   {'type': 'replication',\n",
       "    'numberOfIndividuals': 590,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'European'}],\n",
       "    'countryOfOrigin': [],\n",
       "    'countryOfRecruitment': [{'majorArea': 'Northern America',\n",
       "      'region': None,\n",
       "      'countryName': 'U.S.'}]}],\n",
       "  'diseaseTrait': {'trait': 'HIV-1 progression'},\n",
       "  'genotypingTechnologies': [{'genotypingTechnology': 'Genome-wide genotyping array'}],\n",
       "  'replicationSampleSize': '590 European ancestry seroconverter male cases',\n",
       "  'publicationInfo': {'pubmedId': '20064070',\n",
       "   'publicationDate': '2010-02-15',\n",
       "   'publication': 'J Infect Dis',\n",
       "   'title': 'Multistage genomewide association study identifies a locus at 1q41 associated with rate of HIV-1 disease progression to clinical AIDS.',\n",
       "   'author': {'fullname': 'Herbeck JT', 'orcid': None}},\n",
       "  '_links': {'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000596'},\n",
       "   'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000596{?projection}',\n",
       "    'templated': True},\n",
       "   'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000596/associations?projection=associationByStudy'},\n",
       "   'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000596/snps'},\n",
       "   'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000596/backgroundEfoTraits'},\n",
       "   'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000596/efoTraits'},\n",
       "   'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000596/associations'}}},\n",
       " {'initialSampleSize': '7,827 East Asian ancestry individuals',\n",
       "  'gxe': False,\n",
       "  'gxg': False,\n",
       "  'snpCount': 2500000,\n",
       "  'qualifier': '~',\n",
       "  'imputed': True,\n",
       "  'pooled': False,\n",
       "  'studyDesignComment': None,\n",
       "  'accessionId': 'GCST002233',\n",
       "  'fullPvalueSet': False,\n",
       "  'userRequested': False,\n",
       "  'platforms': [{'manufacturer': 'Affymetrix'}, {'manufacturer': 'Illumina'}],\n",
       "  'ancestries': [{'type': 'initial',\n",
       "    'numberOfIndividuals': 7827,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'East Asian'}],\n",
       "    'countryOfOrigin': [{'majorArea': 'NR',\n",
       "      'region': 'NR',\n",
       "      'countryName': 'NR'}],\n",
       "    'countryOfRecruitment': [{'majorArea': 'Asia',\n",
       "      'region': 'Eastern Asia',\n",
       "      'countryName': 'Republic of Korea'},\n",
       "     {'majorArea': 'Asia',\n",
       "      'region': 'South-eastern Asia',\n",
       "      'countryName': 'Singapore'},\n",
       "     {'majorArea': 'Asia', 'region': 'Eastern Asia', 'countryName': 'China'},\n",
       "     {'majorArea': 'Asia',\n",
       "      'region': 'South-eastern Asia',\n",
       "      'countryName': 'Philippines'}]},\n",
       "   {'type': 'replication',\n",
       "    'numberOfIndividuals': 10250,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'East Asian'}],\n",
       "    'countryOfOrigin': [],\n",
       "    'countryOfRecruitment': [{'majorArea': 'Northern America',\n",
       "      'region': None,\n",
       "      'countryName': 'U.S.'},\n",
       "     {'majorArea': 'Asia',\n",
       "      'region': 'Eastern Asia',\n",
       "      'countryName': 'Republic of Korea'},\n",
       "     {'majorArea': 'Asia', 'region': 'Eastern Asia', 'countryName': 'China'},\n",
       "     {'majorArea': 'Asia',\n",
       "      'region': 'Eastern Asia',\n",
       "      'countryName': 'Japan'}]}],\n",
       "  'diseaseTrait': {'trait': 'Adiponectin levels'},\n",
       "  'genotypingTechnologies': [{'genotypingTechnology': 'Genome-wide genotyping array'}],\n",
       "  'replicationSampleSize': '10,252 East Asian ancestry individuals',\n",
       "  'publicationInfo': {'pubmedId': '24105470',\n",
       "   'publicationDate': '2013-10-14',\n",
       "   'publication': 'Hum Mol Genet',\n",
       "   'title': 'A meta-analysis of genome-wide association studies for adiponectin levels in East Asians identifies a novel locus near WDR11-FGFR2.',\n",
       "   'author': {'fullname': 'Wu Y', 'orcid': None}},\n",
       "  '_links': {'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST002233'},\n",
       "   'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST002233{?projection}',\n",
       "    'templated': True},\n",
       "   'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST002233/associations?projection=associationByStudy'},\n",
       "   'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST002233/associations'},\n",
       "   'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST002233/snps'},\n",
       "   'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST002233/backgroundEfoTraits'},\n",
       "   'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST002233/efoTraits'}}},\n",
       " {'initialSampleSize': '417 European ancestry male individuals, 423 European ancestry female individuals',\n",
       "  'gxe': False,\n",
       "  'gxg': False,\n",
       "  'snpCount': 379319,\n",
       "  'qualifier': None,\n",
       "  'imputed': False,\n",
       "  'pooled': False,\n",
       "  'studyDesignComment': None,\n",
       "  'accessionId': 'GCST000332',\n",
       "  'fullPvalueSet': False,\n",
       "  'userRequested': False,\n",
       "  'platforms': [{'manufacturer': 'Affymetrix'}],\n",
       "  'ancestries': [{'type': 'initial',\n",
       "    'numberOfIndividuals': 840,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'European'}],\n",
       "    'countryOfOrigin': [],\n",
       "    'countryOfRecruitment': [{'majorArea': 'Northern America',\n",
       "      'region': None,\n",
       "      'countryName': 'U.S.'}]},\n",
       "   {'type': 'replication',\n",
       "    'numberOfIndividuals': 7623,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'European'}],\n",
       "    'countryOfOrigin': [],\n",
       "    'countryOfRecruitment': []},\n",
       "   {'type': 'replication',\n",
       "    'numberOfIndividuals': 1251,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'African American or Afro-Caribbean'}],\n",
       "    'countryOfOrigin': [],\n",
       "    'countryOfRecruitment': [{'majorArea': 'Northern America',\n",
       "      'region': None,\n",
       "      'countryName': 'U.S.'}]}],\n",
       "  'diseaseTrait': {'trait': 'Smoking behavior'},\n",
       "  'genotypingTechnologies': [{'genotypingTechnology': 'Genome-wide genotyping array'}],\n",
       "  'replicationSampleSize': '412 African American male individuals and 839 African American female individuals from 402 families, 3,491 European ancestry male individuals and 4,132 European ancestry female individuals from 1,731 families',\n",
       "  'publicationInfo': {'pubmedId': '19188921',\n",
       "   'publicationDate': '2009-02-03',\n",
       "   'publication': 'Mol Psychiatry',\n",
       "   'title': 'Genome-wide association analyses suggested a novel mechanism for smoking behavior regulated by IL15.',\n",
       "   'author': {'fullname': 'Liu YZ', 'orcid': None}},\n",
       "  '_links': {'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000332'},\n",
       "   'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000332{?projection}',\n",
       "    'templated': True},\n",
       "   'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000332/associations?projection=associationByStudy'},\n",
       "   'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000332/snps'},\n",
       "   'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000332/backgroundEfoTraits'},\n",
       "   'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000332/efoTraits'},\n",
       "   'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000332/associations'}}},\n",
       " {'initialSampleSize': '2,247 European ancestry individuals',\n",
       "  'gxe': False,\n",
       "  'gxg': False,\n",
       "  'snpCount': 2500000,\n",
       "  'qualifier': '~',\n",
       "  'imputed': True,\n",
       "  'pooled': False,\n",
       "  'studyDesignComment': None,\n",
       "  'accessionId': 'GCST001848',\n",
       "  'fullPvalueSet': False,\n",
       "  'userRequested': False,\n",
       "  'platforms': [{'manufacturer': 'Illumina'}],\n",
       "  'ancestries': [{'type': 'initial',\n",
       "    'numberOfIndividuals': 2247,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'European'}],\n",
       "    'countryOfOrigin': [],\n",
       "    'countryOfRecruitment': [{'majorArea': 'Europe',\n",
       "      'region': 'Northern Europe',\n",
       "      'countryName': 'Sweden'},\n",
       "     {'majorArea': 'Europe',\n",
       "      'region': 'Northern Europe',\n",
       "      'countryName': 'U.K.'},\n",
       "     {'majorArea': 'Europe',\n",
       "      'region': 'Southern Europe',\n",
       "      'countryName': 'Croatia'}]}],\n",
       "  'diseaseTrait': {'trait': 'IgG glycosylation'},\n",
       "  'genotypingTechnologies': [{'genotypingTechnology': 'Genome-wide genotyping array'}],\n",
       "  'replicationSampleSize': 'NA',\n",
       "  'publicationInfo': {'pubmedId': '23382691',\n",
       "   'publicationDate': '2013-01-31',\n",
       "   'publication': 'PLoS Genet',\n",
       "   'title': 'Loci associated with N-glycosylation of human immunoglobulin G show pleiotropy with autoimmune diseases and haematological cancers.',\n",
       "   'author': {'fullname': 'Lauc G', 'orcid': '0000-0003-1840-9560'}},\n",
       "  '_links': {'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001848'},\n",
       "   'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001848{?projection}',\n",
       "    'templated': True},\n",
       "   'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001848/associations?projection=associationByStudy'},\n",
       "   'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001848/associations'},\n",
       "   'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001848/snps'},\n",
       "   'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001848/backgroundEfoTraits'},\n",
       "   'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001848/efoTraits'}}},\n",
       " {'initialSampleSize': '886 European ancestry individuals',\n",
       "  'gxe': False,\n",
       "  'gxg': False,\n",
       "  'snpCount': 70897,\n",
       "  'qualifier': None,\n",
       "  'imputed': False,\n",
       "  'pooled': False,\n",
       "  'studyDesignComment': None,\n",
       "  'accessionId': 'GCST000082',\n",
       "  'fullPvalueSet': False,\n",
       "  'userRequested': False,\n",
       "  'platforms': [{'manufacturer': 'Affymetrix'}],\n",
       "  'ancestries': [{'type': 'initial',\n",
       "    'numberOfIndividuals': 886,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'European'}],\n",
       "    'countryOfOrigin': [],\n",
       "    'countryOfRecruitment': []}],\n",
       "  'diseaseTrait': {'trait': 'Factor VII'},\n",
       "  'genotypingTechnologies': [{'genotypingTechnology': 'Genome-wide genotyping array'}],\n",
       "  'replicationSampleSize': 'NA',\n",
       "  'publicationInfo': {'pubmedId': '17903294',\n",
       "   'publicationDate': '2007-09-19',\n",
       "   'publication': 'BMC Med Genet',\n",
       "   'title': 'Genome-wide association and linkage analyses of hemostatic factors and hematological phenotypes in the Framingham Heart Study.',\n",
       "   'author': {'fullname': 'Yang Q', 'orcid': '0000-0002-3658-1375'}},\n",
       "  '_links': {'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000082'},\n",
       "   'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000082{?projection}',\n",
       "    'templated': True},\n",
       "   'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000082/associations?projection=associationByStudy'},\n",
       "   'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000082/snps'},\n",
       "   'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000082/backgroundEfoTraits'},\n",
       "   'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000082/efoTraits'},\n",
       "   'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000082/associations'}}},\n",
       " {'initialSampleSize': '1,087 European ancestry individuals from 307 families',\n",
       "  'gxe': False,\n",
       "  'gxg': False,\n",
       "  'snpCount': 66543,\n",
       "  'qualifier': None,\n",
       "  'imputed': False,\n",
       "  'pooled': False,\n",
       "  'studyDesignComment': None,\n",
       "  'accessionId': 'GCST000073',\n",
       "  'fullPvalueSet': False,\n",
       "  'userRequested': False,\n",
       "  'platforms': [{'manufacturer': 'Affymetrix'}],\n",
       "  'ancestries': [{'type': 'initial',\n",
       "    'numberOfIndividuals': 1087,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'European'}],\n",
       "    'countryOfOrigin': [],\n",
       "    'countryOfRecruitment': []},\n",
       "   {'type': 'replication',\n",
       "    'numberOfIndividuals': 4815,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'European'}],\n",
       "    'countryOfOrigin': [],\n",
       "    'countryOfRecruitment': []},\n",
       "   {'type': 'replication',\n",
       "    'numberOfIndividuals': 634,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'Native American'}],\n",
       "    'countryOfOrigin': [],\n",
       "    'countryOfRecruitment': []},\n",
       "   {'type': 'replication',\n",
       "    'numberOfIndividuals': 603,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'Hispanic or Latin American'}],\n",
       "    'countryOfOrigin': [],\n",
       "    'countryOfRecruitment': []}],\n",
       "  'diseaseTrait': {'trait': 'Type 2 diabetes and 6 quantitative traits'},\n",
       "  'genotypingTechnologies': [{'genotypingTechnology': 'Genome-wide genotyping array'}],\n",
       "  'replicationSampleSize': '1,465 European ancestry individuals, 1,464 European ancestry cases, 1,467 European ancestry controls, 300 Pima Indian ancestry cases, 334 Pima Indian ancestry controls, 287 Mexican American cases, 316 Mexican American controls, 124 Old Order Amish cases, 295 Old Order Amish controls',\n",
       "  'publicationInfo': {'pubmedId': '17848626',\n",
       "   'publicationDate': '2007-09-10',\n",
       "   'publication': 'Diabetes',\n",
       "   'title': 'A 100K genome-wide association scan for diabetes and related traits in the Framingham Heart Study: replication and integration with other genome-wide datasets.',\n",
       "   'author': {'fullname': 'Florez JC', 'orcid': None}},\n",
       "  '_links': {'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000073'},\n",
       "   'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000073{?projection}',\n",
       "    'templated': True},\n",
       "   'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000073/associations?projection=associationByStudy'},\n",
       "   'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000073/associations'},\n",
       "   'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000073/snps'},\n",
       "   'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000073/backgroundEfoTraits'},\n",
       "   'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000073/efoTraits'}}}]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import utils\n",
    "res = utils.analyze_json_keywords(metadata_list, [\"diabetes\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "20"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(metadata_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(5.0, 95.0, {'diabetes': 19}, 1, 20, 19)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "______________________________________________________________________________________________________________\n",
      "Principle: F2\n",
      "Description: Findability: Keyword search in metadata\n",
      "Score: 0.05\n",
      "Explanation: 5.00% of entities have all required fields (1 out of 20).\n",
      "95.00% of entities are missing some required fields (19 out of 20).\n",
      "\n",
      "Breakdown of missing fields:\n",
      "19 entities are missing 'suicide' field.\n",
      "______________________________________________________________________________________________________________\n",
      "3,117 European ancestry Bipolar disorder cases, 1,273 European ancestry Major depressive disorder cases\n",
      "37 European ancestry cases, 36 European ancestry controls\n",
      "up to 17,723 European ancestry individuals\n",
      "15,358 European ancestry individuals\n",
      "105 East Asian ancestry cases\n",
      "12,441 European ancestry children\n",
      "2,132 Indian ancestry individuals, 2,313 Malay ancestry individuals\n",
      "1,339 European ancestry cases\n",
      "31,580 European ancestry individuals\n",
      "7,473 African American individuals\n",
      "811 anti-dsDNA positive European ancestry cases, 906 anti-dsDNA negative European ancestry cases, 4,813 European ancestry controls\n",
      "1,831 European ancestry cases, 1,764 European ancestry controls\n",
      "880 Amish individuals, 6,296 European ancestry individuals\n",
      "229 Hispanic individuals from 34 families\n",
      "51 European ancestry rapid progressor male cases, 57 European ancestry moderate progressor male cases, 48 European ancestry long-term progressor male cases\n",
      "7,827 East Asian ancestry individuals\n",
      "417 European ancestry male individuals, 423 European ancestry female individuals\n",
      "2,247 European ancestry individuals\n",
      "886 European ancestry individuals\n",
      "1,087 European ancestry individuals from 307 families\n",
      "______________________________________________________________________________________________________________\n",
      "Principle: F3\n",
      "Description: Findability: Check for identifiers\n",
      "Score: 1.0\n",
      "Explanation: 100.00% of entities have an ID field (20 out of 20).\n",
      "0.00% of entities are missing the ID field (0 out of 20).\n",
      "______________________________________________________________________________________________________________\n",
      "______________________________________________________________________________________________________________\n",
      "Principle: A1\n",
      "Description: Accessibility: Request status\n",
      "Score: 1\n",
      "Explanation: Successful request: API request returned the status 200.\n",
      "______________________________________________________________________________________________________________\n",
      "______________________________________________________________________________________________________________\n",
      "Principle: A1\n",
      "Description: Accessibility: Request status\n",
      "Score: 1\n",
      "Explanation: Successful request: API request returned the status 200.\n",
      "______________________________________________________________________________________________________________\n",
      "______________________________________________________________________________________________________________\n",
      "Principle: A1.1\n",
      "Description: Accessibility: Check after A1\n",
      "Score: 1\n",
      "Explanation: Follow-up check successful: API request returned the status 200.\n",
      "______________________________________________________________________________________________________________\n",
      "______________________________________________________________________________________________________________\n",
      "Principle: A1\n",
      "Description: Accessibility: Request status\n",
      "Score: 1\n",
      "Explanation: Successful request: API request returned the status 200.\n",
      "______________________________________________________________________________________________________________\n",
      "______________________________________________________________________________________________________________\n",
      "Principle: A1.1\n",
      "Description: Accessibility: Check after A1\n",
      "Score: 1\n",
      "Explanation: Follow-up check successful: API request returned the status 200.\n",
      "______________________________________________________________________________________________________________\n",
      "______________________________________________________________________________________________________________\n",
      "Principle: A1.2\n",
      "Description: Accessibility: Another check after A1\n",
      "Score: 1\n",
      "Explanation: Additional check successful: API request returned the status 200.\n",
      "______________________________________________________________________________________________________________\n",
      "(Meta)data is in format: json\n",
      "(Meta)data is in format: json\n",
      "(Meta)data is in format: json\n",
      "(Meta)data is in format: json\n",
      "(Meta)data is in format: json\n",
      "(Meta)data is in format: json\n",
      "(Meta)data is in format: json\n",
      "(Meta)data is in format: json\n",
      "(Meta)data is in format: json\n",
      "(Meta)data is in format: json\n",
      "(Meta)data is in format: json\n",
      "(Meta)data is in format: json\n",
      "(Meta)data is in format: json\n",
      "(Meta)data is in format: json\n",
      "(Meta)data is in format: json\n",
      "(Meta)data is in format: json\n",
      "(Meta)data is in format: json\n",
      "(Meta)data is in format: json\n",
      "(Meta)data is in format: json\n",
      "(Meta)data is in format: json\n",
      "______________________________________________________________________________________________________________\n",
      "Principle: I1\n",
      "Description: Interoperability: Format check\n",
      "Score: 1.0\n",
      "Explanation: Format is json.\n",
      "______________________________________________________________________________________________________________\n",
      "{'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000854'}, 'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000854{?projection}', 'templated': True}, 'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000854/associations?projection=associationByStudy'}, 'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000854/associations'}, 'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000854/snps'}, 'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000854/backgroundEfoTraits'}, 'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000854/efoTraits'}}\n",
      "{'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000841'}, 'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000841{?projection}', 'templated': True}, 'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000841/associations?projection=associationByStudy'}, 'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000841/snps'}, 'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000841/backgroundEfoTraits'}, 'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000841/efoTraits'}, 'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000841/associations'}}\n",
      "{'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000809'}, 'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000809{?projection}', 'templated': True}, 'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000809/associations?projection=associationByStudy'}, 'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000809/associations'}, 'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000809/snps'}, 'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000809/backgroundEfoTraits'}, 'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000809/efoTraits'}}\n",
      "{'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000847'}, 'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000847{?projection}', 'templated': True}, 'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000847/associations?projection=associationByStudy'}, 'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000847/snps'}, 'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000847/backgroundEfoTraits'}, 'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000847/efoTraits'}, 'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000847/associations'}}\n",
      "{'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000871'}, 'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000871{?projection}', 'templated': True}, 'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000871/associations?projection=associationByStudy'}, 'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000871/snps'}, 'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000871/backgroundEfoTraits'}, 'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000871/efoTraits'}, 'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000871/associations'}}\n",
      "{'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001837'}, 'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001837{?projection}', 'templated': True}, 'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001837/associations?projection=associationByStudy'}, 'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001837/associations'}, 'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001837/snps'}, 'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001837/backgroundEfoTraits'}, 'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001837/efoTraits'}}\n",
      "{'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000970'}, 'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000970{?projection}', 'templated': True}, 'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000970/associations?projection=associationByStudy'}, 'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000970/snps'}, 'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000970/backgroundEfoTraits'}, 'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000970/efoTraits'}, 'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000970/associations'}}\n",
      "{'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000981'}, 'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000981{?projection}', 'templated': True}, 'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000981/associations?projection=associationByStudy'}, 'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000981/associations'}, 'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000981/snps'}, 'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000981/backgroundEfoTraits'}, 'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000981/efoTraits'}}\n",
      "{'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000988'}, 'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000988{?projection}', 'templated': True}, 'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000988/associations?projection=associationByStudy'}, 'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000988/snps'}, 'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000988/backgroundEfoTraits'}, 'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000988/efoTraits'}, 'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000988/associations'}}\n",
      "{'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000997'}, 'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000997{?projection}', 'templated': True}, 'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000997/associations?projection=associationByStudy'}, 'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000997/associations'}, 'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000997/snps'}, 'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000997/backgroundEfoTraits'}, 'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000997/efoTraits'}}\n",
      "{'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000996'}, 'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000996{?projection}', 'templated': True}, 'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000996/associations?projection=associationByStudy'}, 'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000996/snps'}, 'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000996/backgroundEfoTraits'}, 'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000996/efoTraits'}, 'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000996/associations'}}\n",
      "{'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000986'}, 'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000986{?projection}', 'templated': True}, 'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000986/associations?projection=associationByStudy'}, 'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000986/associations'}, 'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000986/snps'}, 'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000986/backgroundEfoTraits'}, 'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000986/efoTraits'}}\n",
      "{'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001008'}, 'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001008{?projection}', 'templated': True}, 'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001008/associations?projection=associationByStudy'}, 'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001008/snps'}, 'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001008/backgroundEfoTraits'}, 'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001008/efoTraits'}, 'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001008/associations'}}\n",
      "{'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000711'}, 'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000711{?projection}', 'templated': True}, 'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000711/associations?projection=associationByStudy'}, 'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000711/associations'}, 'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000711/snps'}, 'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000711/backgroundEfoTraits'}, 'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000711/efoTraits'}}\n",
      "{'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000596'}, 'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000596{?projection}', 'templated': True}, 'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000596/associations?projection=associationByStudy'}, 'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000596/snps'}, 'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000596/backgroundEfoTraits'}, 'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000596/efoTraits'}, 'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000596/associations'}}\n",
      "{'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST002233'}, 'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST002233{?projection}', 'templated': True}, 'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST002233/associations?projection=associationByStudy'}, 'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST002233/associations'}, 'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST002233/snps'}, 'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST002233/backgroundEfoTraits'}, 'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST002233/efoTraits'}}\n",
      "{'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000332'}, 'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000332{?projection}', 'templated': True}, 'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000332/associations?projection=associationByStudy'}, 'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000332/snps'}, 'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000332/backgroundEfoTraits'}, 'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000332/efoTraits'}, 'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000332/associations'}}\n",
      "{'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001848'}, 'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001848{?projection}', 'templated': True}, 'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001848/associations?projection=associationByStudy'}, 'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001848/associations'}, 'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001848/snps'}, 'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001848/backgroundEfoTraits'}, 'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001848/efoTraits'}}\n",
      "{'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000082'}, 'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000082{?projection}', 'templated': True}, 'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000082/associations?projection=associationByStudy'}, 'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000082/snps'}, 'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000082/backgroundEfoTraits'}, 'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000082/efoTraits'}, 'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000082/associations'}}\n",
      "{'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000073'}, 'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000073{?projection}', 'templated': True}, 'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000073/associations?projection=associationByStudy'}, 'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000073/associations'}, 'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000073/snps'}, 'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000073/backgroundEfoTraits'}, 'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000073/efoTraits'}}\n",
      "______________________________________________________________________________________________________________\n",
      "Principle: I3\n",
      "Description: Interoperability: Metadata include qualified references to other (meta)data\n",
      "Score: 1.0\n",
      "Explanation: 100.00% of entities have a reference to other metadata.\n",
      "______________________________________________________________________________________________________________\n",
      "None\n",
      "None\n",
      "None\n",
      "None\n",
      "None\n",
      "None\n",
      "None\n",
      "None\n",
      "None\n",
      "None\n",
      "None\n",
      "None\n",
      "None\n",
      "None\n",
      "None\n",
      "None\n",
      "None\n",
      "None\n",
      "None\n",
      "None\n",
      "______________________________________________________________________________________________________________\n",
      "Principle: R1\n",
      "Description: Reusability: Metadata completeness\n",
      "Score: 0.0\n",
      "Explanation: Metadata is incomplete: it does not include the type of object nor the metadata specify the content of the data technical properties of its data file and format.\n",
      "______________________________________________________________________________________________________________\n",
      "______________________________________________________________________________________________________________\n",
      "Principle: R1.1\n",
      "Description: Reusability: Use of community standards\n",
      "Score: 1.0\n",
      "Explanation: All community standards fields present\n",
      "______________________________________________________________________________________________________________\n",
      "{'pubmedId': '21041247', 'publicationDate': '2010-11-01', 'publication': 'Am J Psychiatry', 'title': 'Genome-wide association study of suicide attempts in mood disorder patients.', 'author': {'fullname': 'Perlis RH', 'orcid': None}}\n",
      "{'pubmedId': '20971583', 'publicationDate': '2010-10-22', 'publication': 'J Hand Surg Am', 'title': \"Genome-wide association scan of Dupuytren's disease.\", 'author': {'fullname': 'Ojwang JO', 'orcid': None}}\n",
      "{'pubmedId': '20864672', 'publicationDate': '2010-09-23', 'publication': 'Arterioscler Thromb Vasc Biol', 'title': 'Genetic variants influencing circulating lipid levels and risk of coronary artery disease.', 'author': {'fullname': 'Waterworth DM', 'orcid': None}}\n",
      "{'pubmedId': '21060863', 'publicationDate': '2010-10-28', 'publication': 'PLoS Genet', 'title': 'Four novel Loci (19q13, 6q24, 12q24, and 5q14) influence the microcirculation in vivo.', 'author': {'fullname': 'Ikram MK', 'orcid': None}}\n",
      "{'pubmedId': '21079520', 'publicationDate': '2010-11-12', 'publication': 'J Thorac Oncol', 'title': 'Genome-wide association study on overall survival of advanced non-small cell lung cancer patients treated with carboplatin and paclitaxel.', 'author': {'fullname': 'Sato Y', 'orcid': None}}\n",
      "{'pubmedId': '23358156', 'publicationDate': '2013-01-29', 'publication': 'Mol Psychiatry', 'title': 'Childhood intelligence is heritable, highly polygenic and associated with FNBP1L.', 'author': {'fullname': 'Benyamin B', 'orcid': '0000-0001-5608-2293'}}\n",
      "{'pubmedId': '21307088', 'publicationDate': '2011-02-09', 'publication': 'Hum Mol Genet', 'title': 'Genome-wide association studies in Asians confirm the involvement of ATOH7 and TGFBR3, and further identify CARD10 as a novel locus influencing optic disc area.', 'author': {'fullname': 'Khor CC', 'orcid': '0000-0002-1128-4729'}}\n",
      "{'pubmedId': '21326295', 'publicationDate': '2011-02-17', 'publication': 'J Invest Dermatol', 'title': 'Genome-wide analysis identifies a quantitative trait locus in the MHC class II region associated with generalized vitiligo age of onset.', 'author': {'fullname': 'Jin Y', 'orcid': None}}\n",
      "{'pubmedId': '21355061', 'publicationDate': '2011-02-25', 'publication': 'J Am Soc Nephrol', 'title': 'CUBN is a gene locus for albuminuria.', 'author': {'fullname': 'Bger CA', 'orcid': None}}\n",
      "{'pubmedId': '21378095', 'publicationDate': '2011-03-04', 'publication': 'Hum Mol Genet', 'title': 'Association of genetic variation with systolic and diastolic blood pressure among African Americans: the Candidate Gene Association Resource study.', 'author': {'fullname': 'Fox ER', 'orcid': None}}\n",
      "{'pubmedId': '21408207', 'publicationDate': '2011-03-03', 'publication': 'PLoS Genet', 'title': 'Differential genetic associations for systemic lupus erythematosus based on anti-dsDNA autoantibody production.', 'author': {'fullname': 'Chung SA', 'orcid': None}}\n",
      "{'pubmedId': '21390209', 'publicationDate': '2011-02-24', 'publication': 'PLoS One', 'title': \"Meta-analysis for genome-wide association study identifies multiple variants at the BIN1 locus associated with late-onset Alzheimer's disease.\", 'author': {'fullname': 'Hu X', 'orcid': None}}\n",
      "{'pubmedId': '21423719', 'publicationDate': '2011-03-10', 'publication': 'PLoS Genet', 'title': 'Genome-wide association analysis identifies variants associated with nonalcoholic fatty liver disease that have distinct effects on metabolic traits.', 'author': {'fullname': 'Speliotes EK', 'orcid': None}}\n",
      "{'pubmedId': '20600896', 'publicationDate': '2010-06-26', 'publication': 'J Steroid Biochem Mol Biol', 'title': 'Genome-wide association study of vitamin D concentrations in Hispanic Americans: the IRAS family study.', 'author': {'fullname': 'Engelman CD', 'orcid': None}}\n",
      "{'pubmedId': '20064070', 'publicationDate': '2010-02-15', 'publication': 'J Infect Dis', 'title': 'Multistage genomewide association study identifies a locus at 1q41 associated with rate of HIV-1 disease progression to clinical AIDS.', 'author': {'fullname': 'Herbeck JT', 'orcid': None}}\n",
      "{'pubmedId': '24105470', 'publicationDate': '2013-10-14', 'publication': 'Hum Mol Genet', 'title': 'A meta-analysis of genome-wide association studies for adiponectin levels in East Asians identifies a novel locus near WDR11-FGFR2.', 'author': {'fullname': 'Wu Y', 'orcid': None}}\n",
      "{'pubmedId': '19188921', 'publicationDate': '2009-02-03', 'publication': 'Mol Psychiatry', 'title': 'Genome-wide association analyses suggested a novel mechanism for smoking behavior regulated by IL15.', 'author': {'fullname': 'Liu YZ', 'orcid': None}}\n",
      "{'pubmedId': '23382691', 'publicationDate': '2013-01-31', 'publication': 'PLoS Genet', 'title': 'Loci associated with N-glycosylation of human immunoglobulin G show pleiotropy with autoimmune diseases and haematological cancers.', 'author': {'fullname': 'Lauc G', 'orcid': '0000-0003-1840-9560'}}\n",
      "{'pubmedId': '17903294', 'publicationDate': '2007-09-19', 'publication': 'BMC Med Genet', 'title': 'Genome-wide association and linkage analyses of hemostatic factors and hematological phenotypes in the Framingham Heart Study.', 'author': {'fullname': 'Yang Q', 'orcid': '0000-0002-3658-1375'}}\n",
      "{'pubmedId': '17848626', 'publicationDate': '2007-09-10', 'publication': 'Diabetes', 'title': 'A 100K genome-wide association scan for diabetes and related traits in the Framingham Heart Study: replication and integration with other genome-wide datasets.', 'author': {'fullname': 'Florez JC', 'orcid': None}}\n",
      "______________________________________________________________________________________________________________\n",
      "Principle: R1.2\n",
      "Description: Reusability: Provenance metadata completeness\n",
      "Score: 1.0\n",
      "Explanation: All provenance metadata present\n",
      "______________________________________________________________________________________________________________\n",
      "\n",
      " Metric        Score \n",
      "\n",
      " F2 score       0.05 \n",
      "\n",
      " F3 score       1    \n",
      "\n",
      " A1 score       1    \n",
      "\n",
      " A1.1 score     1    \n",
      "\n",
      " A1.2 score     1    \n",
      "\n",
      " I1 score       1    \n",
      "\n",
      " I3 score       1    \n",
      "\n",
      " R1 score       0    \n",
      "\n",
      " R1.1 score     1    \n",
      "\n",
      " R1.2 score     1    \n",
      "\n"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "'F1 score'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[8], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m scores \u001b[38;5;241m=\u001b[39m \u001b[43massess\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmetadata_list\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkeywords\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrepository_choice\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrequest_status\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/Desktop/ft/FAIR/assess.py:48\u001b[0m, in \u001b[0;36massess\u001b[0;34m(metadata, keywords, repository_choice, url, request_status)\u001b[0m\n\u001b[1;32m     44\u001b[0m     file\u001b[38;5;241m.\u001b[39mwrite(tabulate(data, headers\u001b[38;5;241m=\u001b[39m[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mMetric\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mScore\u001b[39m\u001b[38;5;124m\"\u001b[39m], tablefmt\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mplain\u001b[39m\u001b[38;5;124m\"\u001b[39m))\n\u001b[1;32m     46\u001b[0m scores \u001b[38;5;241m=\u001b[39m {row[\u001b[38;5;241m0\u001b[39m]: row[\u001b[38;5;241m1\u001b[39m] \u001b[38;5;28;01mfor\u001b[39;00m row \u001b[38;5;129;01min\u001b[39;00m data}\n\u001b[0;32m---> 48\u001b[0m \u001b[43mplot_fair_scores\u001b[49m\u001b[43m(\u001b[49m\u001b[43mscores\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     50\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m [row[\u001b[38;5;241m1\u001b[39m] \u001b[38;5;28;01mfor\u001b[39;00m row \u001b[38;5;129;01min\u001b[39;00m data]\n",
      "File \u001b[0;32m~/Desktop/ft/FAIR/assess.py:60\u001b[0m, in \u001b[0;36mplot_fair_scores\u001b[0;34m(scores)\u001b[0m\n\u001b[1;32m     53\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mplot_fair_scores\u001b[39m(scores):\n\u001b[1;32m     54\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m     55\u001b[0m \u001b[38;5;124;03m    Generate a bar graph for FAIR principles scores.\u001b[39;00m\n\u001b[1;32m     56\u001b[0m \n\u001b[1;32m     57\u001b[0m \u001b[38;5;124;03m    Args:\u001b[39;00m\n\u001b[1;32m     58\u001b[0m \u001b[38;5;124;03m        scores (dict): A dictionary of scores for each metric.\u001b[39;00m\n\u001b[1;32m     59\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m---> 60\u001b[0m     F_score \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39mmean([\u001b[43mscores\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mF1 score\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m, scores[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mF2 score\u001b[39m\u001b[38;5;124m'\u001b[39m], scores[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mF3 score\u001b[39m\u001b[38;5;124m'\u001b[39m], scores[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mF4 score\u001b[39m\u001b[38;5;124m'\u001b[39m]])\n\u001b[1;32m     61\u001b[0m     A_score \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39mmean([scores[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mA1 score\u001b[39m\u001b[38;5;124m'\u001b[39m], scores[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mA1.1 score\u001b[39m\u001b[38;5;124m'\u001b[39m], scores[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mA1.2 score\u001b[39m\u001b[38;5;124m'\u001b[39m], scores[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mA2 score\u001b[39m\u001b[38;5;124m'\u001b[39m]])\n\u001b[1;32m     62\u001b[0m     I_score \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39mmean([scores[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mI1 score\u001b[39m\u001b[38;5;124m'\u001b[39m], scores[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mI2 score\u001b[39m\u001b[38;5;124m'\u001b[39m], scores[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mI3 score\u001b[39m\u001b[38;5;124m'\u001b[39m]])\n",
      "\u001b[0;31mKeyError\u001b[0m: 'F1 score'"
     ]
    }
   ],
   "source": [
    "scores = assess(metadata_list, keywords, repository_choice, url, request_status)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
