{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "from repositories import retrieve_metadata\n",
    "from assess import assess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "keywords=[\"melanoma\"]\n",
    "repository_choice=\"5\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetching GDC metadata...\n",
      "Fetching metadata with keywords: ['melanoma']\n"
     ]
    },
    {
     "ename": "SSLError",
     "evalue": "HTTPSConnectionPool(host='api.gdc.cancer.gov', port=443): Max retries exceeded with url: /files?from=0&size=1000&sort=file_size%3Aasc&pretty=true&filters=%7B%22op%22%3A+%22or%22%2C+%22content%22%3A+%5B%7B%22op%22%3A+%22in%22%2C+%22content%22%3A+%7B%22field%22%3A+%22file_name%22%2C+%22value%22%3A+%5B%22melanoma%22%5D%7D%7D%5D%7D (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self-signed certificate in certificate chain (_ssl.c:1007)')))",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mSSLCertVerificationError\u001b[0m                  Traceback (most recent call last)",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/urllib3/connectionpool.py:715\u001b[0m, in \u001b[0;36mHTTPConnectionPool.urlopen\u001b[0;34m(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, **response_kw)\u001b[0m\n\u001b[1;32m    714\u001b[0m \u001b[38;5;66;03m# Make the request on the httplib connection object.\u001b[39;00m\n\u001b[0;32m--> 715\u001b[0m httplib_response \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_make_request\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    716\u001b[0m \u001b[43m    \u001b[49m\u001b[43mconn\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    717\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    718\u001b[0m \u001b[43m    \u001b[49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    719\u001b[0m \u001b[43m    \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeout_obj\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    720\u001b[0m \u001b[43m    \u001b[49m\u001b[43mbody\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mbody\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    721\u001b[0m \u001b[43m    \u001b[49m\u001b[43mheaders\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mheaders\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    722\u001b[0m \u001b[43m    \u001b[49m\u001b[43mchunked\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mchunked\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    723\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    725\u001b[0m \u001b[38;5;66;03m# If we're going to release the connection in ``finally:``, then\u001b[39;00m\n\u001b[1;32m    726\u001b[0m \u001b[38;5;66;03m# the response doesn't need to know about the connection. Otherwise\u001b[39;00m\n\u001b[1;32m    727\u001b[0m \u001b[38;5;66;03m# it will also try to release it and we'll have a double-release\u001b[39;00m\n\u001b[1;32m    728\u001b[0m \u001b[38;5;66;03m# mess.\u001b[39;00m\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/urllib3/connectionpool.py:404\u001b[0m, in \u001b[0;36mHTTPConnectionPool._make_request\u001b[0;34m(self, conn, method, url, timeout, chunked, **httplib_request_kw)\u001b[0m\n\u001b[1;32m    403\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 404\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_validate_conn\u001b[49m\u001b[43m(\u001b[49m\u001b[43mconn\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    405\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m (SocketTimeout, BaseSSLError) \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m    406\u001b[0m     \u001b[38;5;66;03m# Py2 raises this as a BaseSSLError, Py3 raises it as socket timeout.\u001b[39;00m\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/urllib3/connectionpool.py:1058\u001b[0m, in \u001b[0;36mHTTPSConnectionPool._validate_conn\u001b[0;34m(self, conn)\u001b[0m\n\u001b[1;32m   1057\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28mgetattr\u001b[39m(conn, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msock\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m):  \u001b[38;5;66;03m# AppEngine might not have  `.sock`\u001b[39;00m\n\u001b[0;32m-> 1058\u001b[0m     \u001b[43mconn\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mconnect\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1060\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m conn\u001b[38;5;241m.\u001b[39mis_verified:\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/urllib3/connection.py:419\u001b[0m, in \u001b[0;36mHTTPSConnection.connect\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    417\u001b[0m     context\u001b[38;5;241m.\u001b[39mload_default_certs()\n\u001b[0;32m--> 419\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39msock \u001b[38;5;241m=\u001b[39m \u001b[43mssl_wrap_socket\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    420\u001b[0m \u001b[43m    \u001b[49m\u001b[43msock\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mconn\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    421\u001b[0m \u001b[43m    \u001b[49m\u001b[43mkeyfile\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mkey_file\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    422\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcertfile\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcert_file\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    423\u001b[0m \u001b[43m    \u001b[49m\u001b[43mkey_password\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mkey_password\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    424\u001b[0m \u001b[43m    \u001b[49m\u001b[43mca_certs\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mca_certs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    425\u001b[0m \u001b[43m    \u001b[49m\u001b[43mca_cert_dir\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mca_cert_dir\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    426\u001b[0m \u001b[43m    \u001b[49m\u001b[43mca_cert_data\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mca_cert_data\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    427\u001b[0m \u001b[43m    \u001b[49m\u001b[43mserver_hostname\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mserver_hostname\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    428\u001b[0m \u001b[43m    \u001b[49m\u001b[43mssl_context\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcontext\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    429\u001b[0m \u001b[43m    \u001b[49m\u001b[43mtls_in_tls\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtls_in_tls\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    430\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    432\u001b[0m \u001b[38;5;66;03m# If we're using all defaults and the connection\u001b[39;00m\n\u001b[1;32m    433\u001b[0m \u001b[38;5;66;03m# is TLSv1 or TLSv1.1 we throw a DeprecationWarning\u001b[39;00m\n\u001b[1;32m    434\u001b[0m \u001b[38;5;66;03m# for the host.\u001b[39;00m\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/urllib3/util/ssl_.py:449\u001b[0m, in \u001b[0;36mssl_wrap_socket\u001b[0;34m(sock, keyfile, certfile, cert_reqs, ca_certs, server_hostname, ssl_version, ciphers, ssl_context, ca_cert_dir, key_password, ca_cert_data, tls_in_tls)\u001b[0m\n\u001b[1;32m    448\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m send_sni:\n\u001b[0;32m--> 449\u001b[0m     ssl_sock \u001b[38;5;241m=\u001b[39m \u001b[43m_ssl_wrap_socket_impl\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    450\u001b[0m \u001b[43m        \u001b[49m\u001b[43msock\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcontext\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtls_in_tls\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mserver_hostname\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mserver_hostname\u001b[49m\n\u001b[1;32m    451\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    452\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/urllib3/util/ssl_.py:493\u001b[0m, in \u001b[0;36m_ssl_wrap_socket_impl\u001b[0;34m(sock, ssl_context, tls_in_tls, server_hostname)\u001b[0m\n\u001b[1;32m    492\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m server_hostname:\n\u001b[0;32m--> 493\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mssl_context\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mwrap_socket\u001b[49m\u001b[43m(\u001b[49m\u001b[43msock\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mserver_hostname\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mserver_hostname\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    494\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n",
      "File \u001b[0;32m/usr/lib/python3.10/ssl.py:513\u001b[0m, in \u001b[0;36mSSLContext.wrap_socket\u001b[0;34m(self, sock, server_side, do_handshake_on_connect, suppress_ragged_eofs, server_hostname, session)\u001b[0m\n\u001b[1;32m    507\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mwrap_socket\u001b[39m(\u001b[38;5;28mself\u001b[39m, sock, server_side\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m,\n\u001b[1;32m    508\u001b[0m                 do_handshake_on_connect\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m,\n\u001b[1;32m    509\u001b[0m                 suppress_ragged_eofs\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m,\n\u001b[1;32m    510\u001b[0m                 server_hostname\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, session\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m):\n\u001b[1;32m    511\u001b[0m     \u001b[38;5;66;03m# SSLSocket class handles server_hostname encoding before it calls\u001b[39;00m\n\u001b[1;32m    512\u001b[0m     \u001b[38;5;66;03m# ctx._wrap_socket()\u001b[39;00m\n\u001b[0;32m--> 513\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msslsocket_class\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_create\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    514\u001b[0m \u001b[43m        \u001b[49m\u001b[43msock\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msock\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    515\u001b[0m \u001b[43m        \u001b[49m\u001b[43mserver_side\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mserver_side\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    516\u001b[0m \u001b[43m        \u001b[49m\u001b[43mdo_handshake_on_connect\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdo_handshake_on_connect\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    517\u001b[0m \u001b[43m        \u001b[49m\u001b[43msuppress_ragged_eofs\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msuppress_ragged_eofs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    518\u001b[0m \u001b[43m        \u001b[49m\u001b[43mserver_hostname\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mserver_hostname\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    519\u001b[0m \u001b[43m        \u001b[49m\u001b[43mcontext\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    520\u001b[0m \u001b[43m        \u001b[49m\u001b[43msession\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msession\u001b[49m\n\u001b[1;32m    521\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/usr/lib/python3.10/ssl.py:1100\u001b[0m, in \u001b[0;36mSSLSocket._create\u001b[0;34m(cls, sock, server_side, do_handshake_on_connect, suppress_ragged_eofs, server_hostname, context, session)\u001b[0m\n\u001b[1;32m   1099\u001b[0m             \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdo_handshake_on_connect should not be specified for non-blocking sockets\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m-> 1100\u001b[0m         \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdo_handshake\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1101\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m (\u001b[38;5;167;01mOSError\u001b[39;00m, \u001b[38;5;167;01mValueError\u001b[39;00m):\n",
      "File \u001b[0;32m/usr/lib/python3.10/ssl.py:1371\u001b[0m, in \u001b[0;36mSSLSocket.do_handshake\u001b[0;34m(self, block)\u001b[0m\n\u001b[1;32m   1370\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39msettimeout(\u001b[38;5;28;01mNone\u001b[39;00m)\n\u001b[0;32m-> 1371\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_sslobj\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdo_handshake\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1372\u001b[0m \u001b[38;5;28;01mfinally\u001b[39;00m:\n",
      "\u001b[0;31mSSLCertVerificationError\u001b[0m: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self-signed certificate in certificate chain (_ssl.c:1007)",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mMaxRetryError\u001b[0m                             Traceback (most recent call last)",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/requests/adapters.py:440\u001b[0m, in \u001b[0;36mHTTPAdapter.send\u001b[0;34m(self, request, stream, timeout, verify, cert, proxies)\u001b[0m\n\u001b[1;32m    439\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m chunked:\n\u001b[0;32m--> 440\u001b[0m     resp \u001b[38;5;241m=\u001b[39m \u001b[43mconn\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43murlopen\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    441\u001b[0m \u001b[43m        \u001b[49m\u001b[43mmethod\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    442\u001b[0m \u001b[43m        \u001b[49m\u001b[43murl\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    443\u001b[0m \u001b[43m        \u001b[49m\u001b[43mbody\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mbody\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    444\u001b[0m \u001b[43m        \u001b[49m\u001b[43mheaders\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mheaders\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    445\u001b[0m \u001b[43m        \u001b[49m\u001b[43mredirect\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    446\u001b[0m \u001b[43m        \u001b[49m\u001b[43massert_same_host\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    447\u001b[0m \u001b[43m        \u001b[49m\u001b[43mpreload_content\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    448\u001b[0m \u001b[43m        \u001b[49m\u001b[43mdecode_content\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    449\u001b[0m \u001b[43m        \u001b[49m\u001b[43mretries\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmax_retries\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    450\u001b[0m \u001b[43m        \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeout\u001b[49m\n\u001b[1;32m    451\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    453\u001b[0m \u001b[38;5;66;03m# Send the request.\u001b[39;00m\n\u001b[1;32m    454\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/urllib3/connectionpool.py:799\u001b[0m, in \u001b[0;36mHTTPConnectionPool.urlopen\u001b[0;34m(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, **response_kw)\u001b[0m\n\u001b[1;32m    797\u001b[0m     e \u001b[38;5;241m=\u001b[39m ProtocolError(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mConnection aborted.\u001b[39m\u001b[38;5;124m\"\u001b[39m, e)\n\u001b[0;32m--> 799\u001b[0m retries \u001b[38;5;241m=\u001b[39m \u001b[43mretries\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mincrement\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    800\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43merror\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43me\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m_pool\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m_stacktrace\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msys\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mexc_info\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m2\u001b[39;49m\u001b[43m]\u001b[49m\n\u001b[1;32m    801\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    802\u001b[0m retries\u001b[38;5;241m.\u001b[39msleep()\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/urllib3/util/retry.py:592\u001b[0m, in \u001b[0;36mRetry.increment\u001b[0;34m(self, method, url, response, error, _pool, _stacktrace)\u001b[0m\n\u001b[1;32m    591\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m new_retry\u001b[38;5;241m.\u001b[39mis_exhausted():\n\u001b[0;32m--> 592\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m MaxRetryError(_pool, url, error \u001b[38;5;129;01mor\u001b[39;00m ResponseError(cause))\n\u001b[1;32m    594\u001b[0m log\u001b[38;5;241m.\u001b[39mdebug(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mIncremented Retry for (url=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m): \u001b[39m\u001b[38;5;132;01m%r\u001b[39;00m\u001b[38;5;124m\"\u001b[39m, url, new_retry)\n",
      "\u001b[0;31mMaxRetryError\u001b[0m: HTTPSConnectionPool(host='api.gdc.cancer.gov', port=443): Max retries exceeded with url: /files?from=0&size=1000&sort=file_size%3Aasc&pretty=true&filters=%7B%22op%22%3A+%22or%22%2C+%22content%22%3A+%5B%7B%22op%22%3A+%22in%22%2C+%22content%22%3A+%7B%22field%22%3A+%22file_name%22%2C+%22value%22%3A+%5B%22melanoma%22%5D%7D%7D%5D%7D (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self-signed certificate in certificate chain (_ssl.c:1007)')))",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mSSLError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[9], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m metadata_list, request_status, url \u001b[38;5;241m=\u001b[39m \u001b[43mretrieve_metadata\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrepository_choice\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkeywords\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/Desktop/ft/FAIR/repositories.py:38\u001b[0m, in \u001b[0;36mretrieve_metadata\u001b[0;34m(repository_choice, keywords)\u001b[0m\n\u001b[1;32m     36\u001b[0m     metadata_list, request_status, url \u001b[38;5;241m=\u001b[39m fetch_experiments_metadata(keywords)\n\u001b[1;32m     37\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m repository_choice \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m5\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[0;32m---> 38\u001b[0m     metadata_list, request_status, url \u001b[38;5;241m=\u001b[39m \u001b[43mfetch_gdc_metadata\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkeywords\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     39\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m repository_choice \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m6\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[1;32m     40\u001b[0m     metadata_list, request_status, url \u001b[38;5;241m=\u001b[39m fetch_icgc_metadata(keywords)\n",
      "File \u001b[0;32m~/Desktop/ft/FAIR/repositories.py:351\u001b[0m, in \u001b[0;36mfetch_gdc_metadata\u001b[0;34m(keywords)\u001b[0m\n\u001b[1;32m    348\u001b[0m total_files_fetched \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlen\u001b[39m(processed_accessions)\n\u001b[1;32m    350\u001b[0m \u001b[38;5;28;01mwhile\u001b[39;00m \u001b[38;5;28;01mTrue\u001b[39;00m:\n\u001b[0;32m--> 351\u001b[0m     response \u001b[38;5;241m=\u001b[39m \u001b[43mrequests\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[43mbase_url\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mparams\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mparams\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    352\u001b[0m     \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mAPI Call: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mresponse\u001b[38;5;241m.\u001b[39murl\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m    353\u001b[0m     request_status \u001b[38;5;241m=\u001b[39m response\u001b[38;5;241m.\u001b[39mstatus_code\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/requests/api.py:75\u001b[0m, in \u001b[0;36mget\u001b[0;34m(url, params, **kwargs)\u001b[0m\n\u001b[1;32m     64\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mget\u001b[39m(url, params\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[1;32m     65\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124mr\u001b[39m\u001b[38;5;124;03m\"\"\"Sends a GET request.\u001b[39;00m\n\u001b[1;32m     66\u001b[0m \n\u001b[1;32m     67\u001b[0m \u001b[38;5;124;03m    :param url: URL for the new :class:`Request` object.\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     72\u001b[0m \u001b[38;5;124;03m    :rtype: requests.Response\u001b[39;00m\n\u001b[1;32m     73\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m---> 75\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mrequest\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mget\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mparams\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mparams\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/requests/api.py:61\u001b[0m, in \u001b[0;36mrequest\u001b[0;34m(method, url, **kwargs)\u001b[0m\n\u001b[1;32m     57\u001b[0m \u001b[38;5;66;03m# By using the 'with' statement we are sure the session is closed, thus we\u001b[39;00m\n\u001b[1;32m     58\u001b[0m \u001b[38;5;66;03m# avoid leaving sockets open which can trigger a ResourceWarning in some\u001b[39;00m\n\u001b[1;32m     59\u001b[0m \u001b[38;5;66;03m# cases, and look like a memory leak in others.\u001b[39;00m\n\u001b[1;32m     60\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m sessions\u001b[38;5;241m.\u001b[39mSession() \u001b[38;5;28;01mas\u001b[39;00m session:\n\u001b[0;32m---> 61\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43msession\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmethod\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43murl\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/requests/sessions.py:529\u001b[0m, in \u001b[0;36mSession.request\u001b[0;34m(self, method, url, params, data, headers, cookies, files, auth, timeout, allow_redirects, proxies, hooks, stream, verify, cert, json)\u001b[0m\n\u001b[1;32m    524\u001b[0m send_kwargs \u001b[38;5;241m=\u001b[39m {\n\u001b[1;32m    525\u001b[0m     \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mtimeout\u001b[39m\u001b[38;5;124m'\u001b[39m: timeout,\n\u001b[1;32m    526\u001b[0m     \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mallow_redirects\u001b[39m\u001b[38;5;124m'\u001b[39m: allow_redirects,\n\u001b[1;32m    527\u001b[0m }\n\u001b[1;32m    528\u001b[0m send_kwargs\u001b[38;5;241m.\u001b[39mupdate(settings)\n\u001b[0;32m--> 529\u001b[0m resp \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msend\u001b[49m\u001b[43m(\u001b[49m\u001b[43mprep\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43msend_kwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    531\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m resp\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/requests/sessions.py:645\u001b[0m, in \u001b[0;36mSession.send\u001b[0;34m(self, request, **kwargs)\u001b[0m\n\u001b[1;32m    642\u001b[0m start \u001b[38;5;241m=\u001b[39m preferred_clock()\n\u001b[1;32m    644\u001b[0m \u001b[38;5;66;03m# Send the request\u001b[39;00m\n\u001b[0;32m--> 645\u001b[0m r \u001b[38;5;241m=\u001b[39m \u001b[43madapter\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msend\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrequest\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    647\u001b[0m \u001b[38;5;66;03m# Total elapsed time of the request (approximately)\u001b[39;00m\n\u001b[1;32m    648\u001b[0m elapsed \u001b[38;5;241m=\u001b[39m preferred_clock() \u001b[38;5;241m-\u001b[39m start\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/requests/adapters.py:517\u001b[0m, in \u001b[0;36mHTTPAdapter.send\u001b[0;34m(self, request, stream, timeout, verify, cert, proxies)\u001b[0m\n\u001b[1;32m    513\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m ProxyError(e, request\u001b[38;5;241m=\u001b[39mrequest)\n\u001b[1;32m    515\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(e\u001b[38;5;241m.\u001b[39mreason, _SSLError):\n\u001b[1;32m    516\u001b[0m         \u001b[38;5;66;03m# This branch is for urllib3 v1.22 and later.\u001b[39;00m\n\u001b[0;32m--> 517\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m SSLError(e, request\u001b[38;5;241m=\u001b[39mrequest)\n\u001b[1;32m    519\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mConnectionError\u001b[39;00m(e, request\u001b[38;5;241m=\u001b[39mrequest)\n\u001b[1;32m    521\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m ClosedPoolError \u001b[38;5;28;01mas\u001b[39;00m e:\n",
      "\u001b[0;31mSSLError\u001b[0m: HTTPSConnectionPool(host='api.gdc.cancer.gov', port=443): Max retries exceeded with url: /files?from=0&size=1000&sort=file_size%3Aasc&pretty=true&filters=%7B%22op%22%3A+%22or%22%2C+%22content%22%3A+%5B%7B%22op%22%3A+%22in%22%2C+%22content%22%3A+%7B%22field%22%3A+%22file_name%22%2C+%22value%22%3A+%5B%22melanoma%22%5D%7D%7D%5D%7D (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self-signed certificate in certificate chain (_ssl.c:1007)')))"
     ]
    }
   ],
   "source": [
    "metadata_list, request_status, url = retrieve_metadata(repository_choice, keywords)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'initialSampleSize': '864 European ancestry cases, 864 European ancestry controls',\n",
       "  'gxe': False,\n",
       "  'gxg': False,\n",
       "  'snpCount': 535150,\n",
       "  'qualifier': None,\n",
       "  'imputed': False,\n",
       "  'pooled': True,\n",
       "  'studyDesignComment': None,\n",
       "  'accessionId': 'GCST000198',\n",
       "  'fullPvalueSet': False,\n",
       "  'userRequested': False,\n",
       "  'platforms': [{'manufacturer': 'Illumina'}],\n",
       "  'ancestries': [{'type': 'initial',\n",
       "    'numberOfIndividuals': 1728,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'European'}],\n",
       "    'countryOfOrigin': [],\n",
       "    'countryOfRecruitment': [{'majorArea': 'Oceania',\n",
       "      'region': 'Australia/New Zealand',\n",
       "      'countryName': 'Australia'}]},\n",
       "   {'type': 'replication',\n",
       "    'numberOfIndividuals': 2481,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'European'}],\n",
       "    'countryOfOrigin': [],\n",
       "    'countryOfRecruitment': [{'majorArea': 'Oceania',\n",
       "      'region': 'Australia/New Zealand',\n",
       "      'countryName': 'Australia'}]}],\n",
       "  'diseaseTrait': {'trait': 'Melanoma'},\n",
       "  'genotypingTechnologies': [{'genotypingTechnology': 'Genome-wide genotyping array'}],\n",
       "  'replicationSampleSize': '1,230 European ancestry cases, 1,251 European ancestry controls',\n",
       "  'publicationInfo': {'pubmedId': '18488026',\n",
       "   'publicationDate': '2008-05-18',\n",
       "   'publication': 'Nat Genet',\n",
       "   'title': 'Common sequence variants on 20q11.22 confer melanoma susceptibility.',\n",
       "   'author': {'fullname': 'Brown KM', 'orcid': None}},\n",
       "  '_links': {'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000198'},\n",
       "   'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000198{?projection}',\n",
       "    'templated': True},\n",
       "   'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000198/associations?projection=associationByStudy'},\n",
       "   'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000198/associations'},\n",
       "   'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000198/snps'},\n",
       "   'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000198/backgroundEfoTraits'},\n",
       "   'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000198/efoTraits'}}},\n",
       " {'initialSampleSize': '1,539 European ancestry cases, 3,917 European ancestry controls',\n",
       "  'gxe': False,\n",
       "  'gxg': False,\n",
       "  'snpCount': 317000,\n",
       "  'qualifier': '~',\n",
       "  'imputed': False,\n",
       "  'pooled': False,\n",
       "  'studyDesignComment': None,\n",
       "  'accessionId': 'GCST000437',\n",
       "  'fullPvalueSet': False,\n",
       "  'userRequested': False,\n",
       "  'platforms': [{'manufacturer': 'Illumina'}],\n",
       "  'ancestries': [{'type': 'initial',\n",
       "    'numberOfIndividuals': 5456,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'European'}],\n",
       "    'countryOfOrigin': [],\n",
       "    'countryOfRecruitment': [{'majorArea': 'Europe',\n",
       "      'region': 'Northern Europe',\n",
       "      'countryName': 'Sweden'},\n",
       "     {'majorArea': 'Oceania',\n",
       "      'region': 'Australia/New Zealand',\n",
       "      'countryName': 'Australia'},\n",
       "     {'majorArea': 'Europe',\n",
       "      'region': 'Southern Europe',\n",
       "      'countryName': 'Italy'},\n",
       "     {'majorArea': 'Europe',\n",
       "      'region': 'Western Europe',\n",
       "      'countryName': 'Netherlands'},\n",
       "     {'majorArea': 'Europe',\n",
       "      'region': 'Northern Europe',\n",
       "      'countryName': 'U.K.'},\n",
       "     {'majorArea': 'Europe',\n",
       "      'region': 'Southern Europe',\n",
       "      'countryName': 'Spain'},\n",
       "     {'majorArea': 'Europe',\n",
       "      'region': 'Western Europe',\n",
       "      'countryName': 'France'}]},\n",
       "   {'type': 'replication',\n",
       "    'numberOfIndividuals': 4179,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'European'}],\n",
       "    'countryOfOrigin': [],\n",
       "    'countryOfRecruitment': [{'majorArea': 'Europe',\n",
       "      'region': 'Northern Europe',\n",
       "      'countryName': 'Sweden'},\n",
       "     {'majorArea': 'Asia', 'region': 'Western Asia', 'countryName': 'Israel'},\n",
       "     {'majorArea': 'Oceania',\n",
       "      'region': 'Australia/New Zealand',\n",
       "      'countryName': 'Australia'},\n",
       "     {'majorArea': 'Europe',\n",
       "      'region': 'Southern Europe',\n",
       "      'countryName': 'Italy'},\n",
       "     {'majorArea': 'Europe',\n",
       "      'region': 'Western Europe',\n",
       "      'countryName': 'Netherlands'},\n",
       "     {'majorArea': 'Europe',\n",
       "      'region': 'Northern Europe',\n",
       "      'countryName': 'U.K.'},\n",
       "     {'majorArea': 'Europe',\n",
       "      'region': 'Western Europe',\n",
       "      'countryName': 'France'}]}],\n",
       "  'diseaseTrait': {'trait': 'Melanoma'},\n",
       "  'genotypingTechnologies': [{'genotypingTechnology': 'Genome-wide genotyping array'}],\n",
       "  'replicationSampleSize': '2,312 European ancestry cases, 1,867 European ancestry controls',\n",
       "  'publicationInfo': {'pubmedId': '19578364',\n",
       "   'publicationDate': '2009-07-05',\n",
       "   'publication': 'Nat Genet',\n",
       "   'title': 'Genome-wide association study identifies three loci associated with melanoma risk.',\n",
       "   'author': {'fullname': 'Bishop DT', 'orcid': None}},\n",
       "  '_links': {'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000437'},\n",
       "   'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000437{?projection}',\n",
       "    'templated': True},\n",
       "   'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000437/associations?projection=associationByStudy'},\n",
       "   'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000437/snps'},\n",
       "   'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000437/backgroundEfoTraits'},\n",
       "   'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000437/efoTraits'},\n",
       "   'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000437/associations'}}},\n",
       " {'initialSampleSize': '1,353 European ancestry cases, 3,566 European ancestry controls',\n",
       "  'gxe': False,\n",
       "  'gxg': False,\n",
       "  'snpCount': 2600000,\n",
       "  'qualifier': None,\n",
       "  'imputed': True,\n",
       "  'pooled': False,\n",
       "  'studyDesignComment': None,\n",
       "  'accessionId': 'GCST001886',\n",
       "  'fullPvalueSet': False,\n",
       "  'userRequested': False,\n",
       "  'platforms': [{'manufacturer': 'Illumina'}],\n",
       "  'ancestries': [{'type': 'initial',\n",
       "    'numberOfIndividuals': 4919,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'European'}],\n",
       "    'countryOfOrigin': [],\n",
       "    'countryOfRecruitment': []},\n",
       "   {'type': 'replication',\n",
       "    'numberOfIndividuals': 67980,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'European'}],\n",
       "    'countryOfOrigin': [],\n",
       "    'countryOfRecruitment': [{'majorArea': 'Europe',\n",
       "      'region': 'Northern Europe',\n",
       "      'countryName': 'Sweden'},\n",
       "     {'majorArea': 'Northern America', 'region': None, 'countryName': 'U.S.'},\n",
       "     {'majorArea': 'Oceania',\n",
       "      'region': 'Australia/New Zealand',\n",
       "      'countryName': 'Australia'},\n",
       "     {'majorArea': 'Europe',\n",
       "      'region': 'Northern Europe',\n",
       "      'countryName': 'Iceland'},\n",
       "     {'majorArea': 'Europe',\n",
       "      'region': 'Southern Europe',\n",
       "      'countryName': 'Italy'},\n",
       "     {'majorArea': 'Europe',\n",
       "      'region': 'Western Europe',\n",
       "      'countryName': 'Netherlands'},\n",
       "     {'majorArea': 'Europe',\n",
       "      'region': 'Southern Europe',\n",
       "      'countryName': 'Spain'},\n",
       "     {'majorArea': 'Europe',\n",
       "      'region': 'Western Europe',\n",
       "      'countryName': 'Austria'}]}],\n",
       "  'diseaseTrait': {'trait': 'Melanoma'},\n",
       "  'genotypingTechnologies': [{'genotypingTechnology': 'Genome-wide genotyping array'}],\n",
       "  'replicationSampleSize': '12,313 European ancestry cases, Up to 55,667 European ancestry controls',\n",
       "  'publicationInfo': {'pubmedId': '23455637',\n",
       "   'publicationDate': '2013-03-03',\n",
       "   'publication': 'Nat Genet',\n",
       "   'title': 'A variant in FTO shows association with melanoma risk not due to BMI.',\n",
       "   'author': {'fullname': 'Iles MM', 'orcid': None}},\n",
       "  '_links': {'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001886'},\n",
       "   'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001886{?projection}',\n",
       "    'templated': True},\n",
       "   'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001886/associations?projection=associationByStudy'},\n",
       "   'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001886/associations'},\n",
       "   'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001886/snps'},\n",
       "   'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001886/backgroundEfoTraits'},\n",
       "   'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001886/efoTraits'}}},\n",
       " {'initialSampleSize': '156 European ancestry cases from 34 high-risk families, 2,150 European ancestry controls',\n",
       "  'gxe': False,\n",
       "  'gxg': False,\n",
       "  'snpCount': 491227,\n",
       "  'qualifier': None,\n",
       "  'imputed': False,\n",
       "  'pooled': False,\n",
       "  'studyDesignComment': None,\n",
       "  'accessionId': 'GCST001129',\n",
       "  'fullPvalueSet': False,\n",
       "  'userRequested': False,\n",
       "  'platforms': [{'manufacturer': 'Illumina'}],\n",
       "  'ancestries': [{'type': 'initial',\n",
       "    'numberOfIndividuals': 2306,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'European'}],\n",
       "    'countryOfOrigin': [],\n",
       "    'countryOfRecruitment': [{'majorArea': 'Northern America',\n",
       "      'region': None,\n",
       "      'countryName': 'U.S.'}]}],\n",
       "  'diseaseTrait': {'trait': 'Melanoma'},\n",
       "  'genotypingTechnologies': [{'genotypingTechnology': 'Genome-wide genotyping array'}],\n",
       "  'replicationSampleSize': 'NA',\n",
       "  'publicationInfo': {'pubmedId': '21706340',\n",
       "   'publicationDate': '2011-06-26',\n",
       "   'publication': 'Hum Genet',\n",
       "   'title': 'A unique genome-wide association analysis in extended Utah high-risk pedigrees identifies a novel melanoma risk variant on chromosome arm 10q.',\n",
       "   'author': {'fullname': 'Teerlink C', 'orcid': '0000-0002-1992-2326'}},\n",
       "  '_links': {'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001129'},\n",
       "   'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001129{?projection}',\n",
       "    'templated': True},\n",
       "   'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001129/associations?projection=associationByStudy'},\n",
       "   'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001129/snps'},\n",
       "   'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001129/backgroundEfoTraits'},\n",
       "   'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001129/efoTraits'},\n",
       "   'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001129/associations'}}},\n",
       " {'initialSampleSize': '2,168 European ancestry cases, 4,387 European ancestry controls',\n",
       "  'gxe': False,\n",
       "  'gxg': False,\n",
       "  'snpCount': 5480804,\n",
       "  'qualifier': None,\n",
       "  'imputed': True,\n",
       "  'pooled': False,\n",
       "  'studyDesignComment': None,\n",
       "  'accessionId': 'GCST001266',\n",
       "  'fullPvalueSet': False,\n",
       "  'userRequested': False,\n",
       "  'platforms': [{'manufacturer': 'Illumina'}],\n",
       "  'ancestries': [{'type': 'initial',\n",
       "    'numberOfIndividuals': 6555,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'European'}],\n",
       "    'countryOfOrigin': [],\n",
       "    'countryOfRecruitment': [{'majorArea': 'Oceania',\n",
       "      'region': 'Australia/New Zealand',\n",
       "      'countryName': 'Australia'}]},\n",
       "   {'type': 'replication',\n",
       "    'numberOfIndividuals': 20337,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'European'}],\n",
       "    'countryOfOrigin': [],\n",
       "    'countryOfRecruitment': [{'majorArea': 'Northern America',\n",
       "      'region': None,\n",
       "      'countryName': 'U.S.'},\n",
       "     {'majorArea': 'Asia',\n",
       "      'region': 'Western Asia',\n",
       "      'countryName': 'Israel'}]}],\n",
       "  'diseaseTrait': {'trait': 'Melanoma'},\n",
       "  'genotypingTechnologies': [{'genotypingTechnology': 'Genome-wide genotyping array'}],\n",
       "  'replicationSampleSize': '5,193 European ancestry cases, 15,144 European ancestry controls',\n",
       "  'publicationInfo': {'pubmedId': '21983785',\n",
       "   'publicationDate': '2011-10-09',\n",
       "   'publication': 'Nat Genet',\n",
       "   'title': 'Genome-wide association study identifies a new melanoma susceptibility locus at 1q21.3.',\n",
       "   'author': {'fullname': 'Macgregor S', 'orcid': '0000-0001-6731-8142'}},\n",
       "  '_links': {'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001266'},\n",
       "   'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001266{?projection}',\n",
       "    'templated': True},\n",
       "   'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001266/associations?projection=associationByStudy'},\n",
       "   'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001266/associations'},\n",
       "   'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001266/snps'},\n",
       "   'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001266/backgroundEfoTraits'},\n",
       "   'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001266/efoTraits'}}},\n",
       " {'initialSampleSize': '2,804 European ancestry cases, 7,618 European ancestry controls',\n",
       "  'gxe': False,\n",
       "  'gxg': False,\n",
       "  'snpCount': 594997,\n",
       "  'qualifier': None,\n",
       "  'imputed': False,\n",
       "  'pooled': False,\n",
       "  'studyDesignComment': None,\n",
       "  'accessionId': 'GCST001267',\n",
       "  'fullPvalueSet': False,\n",
       "  'userRequested': False,\n",
       "  'platforms': [{'manufacturer': 'Illumina'}],\n",
       "  'ancestries': [{'type': 'replication',\n",
       "    'numberOfIndividuals': 13000,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'European'}],\n",
       "    'countryOfOrigin': [],\n",
       "    'countryOfRecruitment': [{'majorArea': 'Northern America',\n",
       "      'region': None,\n",
       "      'countryName': 'U.S.'},\n",
       "     {'majorArea': 'Oceania',\n",
       "      'region': 'Australia/New Zealand',\n",
       "      'countryName': 'Australia'},\n",
       "     {'majorArea': 'Europe',\n",
       "      'region': 'Western Europe',\n",
       "      'countryName': 'Netherlands'},\n",
       "     {'majorArea': 'Europe',\n",
       "      'region': 'Northern Europe',\n",
       "      'countryName': 'U.K.'}]},\n",
       "   {'type': 'initial',\n",
       "    'numberOfIndividuals': 10422,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'European'}],\n",
       "    'countryOfOrigin': [],\n",
       "    'countryOfRecruitment': [{'majorArea': 'Europe',\n",
       "      'region': 'Northern Europe',\n",
       "      'countryName': 'Sweden'},\n",
       "     {'majorArea': 'Northern America', 'region': None, 'countryName': 'U.S.'},\n",
       "     {'majorArea': 'Europe',\n",
       "      'region': 'Eastern Europe',\n",
       "      'countryName': 'Poland'},\n",
       "     {'majorArea': 'Asia', 'region': 'Western Asia', 'countryName': 'Israel'},\n",
       "     {'majorArea': 'Oceania',\n",
       "      'region': 'Australia/New Zealand',\n",
       "      'countryName': 'Australia'},\n",
       "     {'majorArea': 'Europe',\n",
       "      'region': 'Southern Europe',\n",
       "      'countryName': 'Italy'},\n",
       "     {'majorArea': 'Europe',\n",
       "      'region': 'Western Europe',\n",
       "      'countryName': 'Netherlands'},\n",
       "     {'majorArea': 'Europe',\n",
       "      'region': 'Northern Europe',\n",
       "      'countryName': 'U.K.'},\n",
       "     {'majorArea': 'Europe',\n",
       "      'region': 'Southern Europe',\n",
       "      'countryName': 'Spain'},\n",
       "     {'majorArea': 'Europe',\n",
       "      'region': 'Western Europe',\n",
       "      'countryName': 'France'},\n",
       "     {'majorArea': 'Europe',\n",
       "      'region': 'Northern Europe',\n",
       "      'countryName': 'Norway'}]}],\n",
       "  'diseaseTrait': {'trait': 'Melanoma'},\n",
       "  'genotypingTechnologies': [{'genotypingTechnology': 'Genome-wide genotyping array'}],\n",
       "  'replicationSampleSize': '5,551 European ancestry cases, 7,449 European ancestry controls',\n",
       "  'publicationInfo': {'pubmedId': '21983787',\n",
       "   'publicationDate': '2011-10-09',\n",
       "   'publication': 'Nat Genet',\n",
       "   'title': 'Genome-wide association study identifies three new melanoma susceptibility loci.',\n",
       "   'author': {'fullname': 'Barrett JH', 'orcid': None}},\n",
       "  '_links': {'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001267'},\n",
       "   'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001267{?projection}',\n",
       "    'templated': True},\n",
       "   'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001267/associations?projection=associationByStudy'},\n",
       "   'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001267/snps'},\n",
       "   'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001267/backgroundEfoTraits'},\n",
       "   'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001267/efoTraits'},\n",
       "   'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001267/associations'}}},\n",
       " {'initialSampleSize': '1,804 European ancestry cases, 1,026 European ancestry controls',\n",
       "  'gxe': False,\n",
       "  'gxg': False,\n",
       "  'snpCount': 818977,\n",
       "  'qualifier': None,\n",
       "  'imputed': False,\n",
       "  'pooled': False,\n",
       "  'studyDesignComment': None,\n",
       "  'accessionId': 'GCST001245',\n",
       "  'fullPvalueSet': False,\n",
       "  'userRequested': False,\n",
       "  'platforms': [{'manufacturer': 'Illumina'}],\n",
       "  'ancestries': [{'type': 'initial',\n",
       "    'numberOfIndividuals': 2830,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'European'}],\n",
       "    'countryOfOrigin': [],\n",
       "    'countryOfRecruitment': [{'majorArea': 'Northern America',\n",
       "      'region': None,\n",
       "      'countryName': 'U.S.'}]},\n",
       "   {'type': 'replication',\n",
       "    'numberOfIndividuals': 29807,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'European'},\n",
       "     {'ancestralGroup': 'Greater Middle Eastern (Middle Eastern, North African or Persian)'}],\n",
       "    'countryOfOrigin': [],\n",
       "    'countryOfRecruitment': [{'majorArea': 'Northern America',\n",
       "      'region': None,\n",
       "      'countryName': 'U.S.'},\n",
       "     {'majorArea': 'Asia', 'region': 'Western Asia', 'countryName': 'Israel'},\n",
       "     {'majorArea': 'Oceania',\n",
       "      'region': 'Australia/New Zealand',\n",
       "      'countryName': 'Australia'},\n",
       "     {'majorArea': 'Europe',\n",
       "      'region': 'Western Europe',\n",
       "      'countryName': 'France'}]}],\n",
       "  'diseaseTrait': {'trait': 'Melanoma'},\n",
       "  'genotypingTechnologies': [{'genotypingTechnology': 'Genome-wide genotyping array'}],\n",
       "  'replicationSampleSize': '6,483 European and Middle East/North African ancestry cases, 23,324 European and Middle East/North African ancestry controls',\n",
       "  'publicationInfo': {'pubmedId': '21926416',\n",
       "   'publicationDate': '2011-09-22',\n",
       "   'publication': 'Hum Mol Genet',\n",
       "   'title': 'Genome-wide association study identifies novel loci predisposing to cutaneous melanoma.',\n",
       "   'author': {'fullname': 'Amos CI', 'orcid': None}},\n",
       "  '_links': {'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001245'},\n",
       "   'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001245{?projection}',\n",
       "    'templated': True},\n",
       "   'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001245/associations?projection=associationByStudy'},\n",
       "   'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001245/associations'},\n",
       "   'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001245/snps'},\n",
       "   'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001245/backgroundEfoTraits'},\n",
       "   'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001245/efoTraits'}}},\n",
       " {'initialSampleSize': '494 European ancestry cases, 5,628 European ancestry controls',\n",
       "  'gxe': False,\n",
       "  'gxg': False,\n",
       "  'snpCount': 1579307,\n",
       "  'qualifier': None,\n",
       "  'imputed': True,\n",
       "  'pooled': False,\n",
       "  'studyDesignComment': None,\n",
       "  'accessionId': 'GCST002514',\n",
       "  'fullPvalueSet': False,\n",
       "  'userRequested': False,\n",
       "  'platforms': [{'manufacturer': 'Affymetrix'}, {'manufacturer': 'Illumina'}],\n",
       "  'ancestries': [{'type': 'replication',\n",
       "    'numberOfIndividuals': 18416,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'European'}],\n",
       "    'countryOfOrigin': [],\n",
       "    'countryOfRecruitment': [{'majorArea': 'Northern America',\n",
       "      'region': None,\n",
       "      'countryName': 'U.S.'},\n",
       "     {'majorArea': 'Asia', 'region': 'Western Asia', 'countryName': 'Israel'},\n",
       "     {'majorArea': 'Oceania',\n",
       "      'region': 'Australia/New Zealand',\n",
       "      'countryName': 'Australia'}]},\n",
       "   {'type': 'initial',\n",
       "    'numberOfIndividuals': 6122,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'European'}],\n",
       "    'countryOfOrigin': [],\n",
       "    'countryOfRecruitment': [{'majorArea': 'Northern America',\n",
       "      'region': None,\n",
       "      'countryName': 'U.S.'}]}],\n",
       "  'diseaseTrait': {'trait': 'Melanoma'},\n",
       "  'genotypingTechnologies': [{'genotypingTechnology': 'Genome-wide genotyping array'}],\n",
       "  'replicationSampleSize': '5,383 European ancestry cases, 13,033 European ancestry controls',\n",
       "  'publicationInfo': {'pubmedId': '24980573',\n",
       "   'publicationDate': '2014-06-30',\n",
       "   'publication': 'Carcinogenesis',\n",
       "   'title': 'Identification of a melanoma susceptibility locus and somatic mutation in TET2.',\n",
       "   'author': {'fullname': 'Song F', 'orcid': None}},\n",
       "  '_links': {'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST002514'},\n",
       "   'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST002514{?projection}',\n",
       "    'templated': True},\n",
       "   'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST002514/associations?projection=associationByStudy'},\n",
       "   'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST002514/associations'},\n",
       "   'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST002514/snps'},\n",
       "   'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST002514/backgroundEfoTraits'},\n",
       "   'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST002514/efoTraits'}}},\n",
       " {'initialSampleSize': '4,842 European ancestry cases, 286,565 European ancestry controls',\n",
       "  'gxe': False,\n",
       "  'gxg': False,\n",
       "  'snpCount': 560000,\n",
       "  'qualifier': 'at least',\n",
       "  'imputed': True,\n",
       "  'pooled': False,\n",
       "  'studyDesignComment': None,\n",
       "  'accessionId': 'GCST004142',\n",
       "  'fullPvalueSet': False,\n",
       "  'userRequested': False,\n",
       "  'platforms': [{'manufacturer': 'Illumina'}],\n",
       "  'ancestries': [{'type': 'initial',\n",
       "    'numberOfIndividuals': 291407,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'European'}],\n",
       "    'countryOfOrigin': [{'majorArea': 'NR',\n",
       "      'region': 'NR',\n",
       "      'countryName': 'NR'}],\n",
       "    'countryOfRecruitment': [{'majorArea': 'Northern America',\n",
       "      'region': None,\n",
       "      'countryName': 'U.S.'}]},\n",
       "   {'type': 'replication',\n",
       "    'numberOfIndividuals': 2830,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'European'}],\n",
       "    'countryOfOrigin': [{'majorArea': 'NR',\n",
       "      'region': 'NR',\n",
       "      'countryName': 'NR'}],\n",
       "    'countryOfRecruitment': [{'majorArea': 'Northern America',\n",
       "      'region': None,\n",
       "      'countryName': 'U.S.'}]}],\n",
       "  'diseaseTrait': {'trait': 'Melanoma'},\n",
       "  'genotypingTechnologies': [{'genotypingTechnology': 'Genome-wide genotyping array'}],\n",
       "  'replicationSampleSize': '1,804 European ancestry cases, 1,026 European ancestry controls',\n",
       "  'publicationInfo': {'pubmedId': '28212542',\n",
       "   'publicationDate': '2017-02-09',\n",
       "   'publication': 'Oncotarget',\n",
       "   'title': 'Two-stage genome-wide association study identifies a novel susceptibility locus associated with melanoma.',\n",
       "   'author': {'fullname': 'Ransohoff KJ', 'orcid': None}},\n",
       "  '_links': {'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST004142'},\n",
       "   'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST004142{?projection}',\n",
       "    'templated': True},\n",
       "   'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST004142/associations?projection=associationByStudy'},\n",
       "   'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST004142/associations'},\n",
       "   'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST004142/snps'},\n",
       "   'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST004142/backgroundEfoTraits'},\n",
       "   'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST004142/efoTraits'}}},\n",
       " {'cohort': 'UKB|GERA',\n",
       "  'initialSampleSize': '6,777 European ancestry cases, 410,350 European ancestry controls',\n",
       "  'gxe': False,\n",
       "  'gxg': False,\n",
       "  'snpCount': 7846216,\n",
       "  'qualifier': None,\n",
       "  'imputed': True,\n",
       "  'pooled': False,\n",
       "  'studyDesignComment': None,\n",
       "  'accessionId': 'GCST90011809',\n",
       "  'fullPvalueSet': True,\n",
       "  'userRequested': False,\n",
       "  'platforms': [{'manufacturer': 'Affymetrix'}],\n",
       "  'ancestries': [{'type': 'initial',\n",
       "    'numberOfIndividuals': 417127,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'European'}],\n",
       "    'countryOfOrigin': [],\n",
       "    'countryOfRecruitment': [{'majorArea': 'Northern America',\n",
       "      'region': None,\n",
       "      'countryName': 'U.S.'},\n",
       "     {'majorArea': 'Europe',\n",
       "      'region': 'Northern Europe',\n",
       "      'countryName': 'U.K.'}]}],\n",
       "  'diseaseTrait': {'trait': 'Melanoma'},\n",
       "  'genotypingTechnologies': [{'genotypingTechnology': 'Genome-wide genotyping array'}],\n",
       "  'replicationSampleSize': 'NA',\n",
       "  'publicationInfo': {'pubmedId': '32887889',\n",
       "   'publicationDate': '2020-09-04',\n",
       "   'publication': 'Nat Commun',\n",
       "   'title': 'Pan-cancer study detects genetic risk variants and shared genetic basis in two large cohorts.',\n",
       "   'author': {'fullname': 'Rashkin SR', 'orcid': None}},\n",
       "  '_links': {'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST90011809'},\n",
       "   'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST90011809{?projection}',\n",
       "    'templated': True},\n",
       "   'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST90011809/associations?projection=associationByStudy'},\n",
       "   'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST90011809/associations'},\n",
       "   'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST90011809/snps'},\n",
       "   'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST90011809/backgroundEfoTraits'},\n",
       "   'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST90011809/efoTraits'}}},\n",
       " {'cohort': 'MIA',\n",
       "  'initialSampleSize': '4,089 European ancestry cases, 11,901 European ancestry controls',\n",
       "  'gxe': False,\n",
       "  'gxg': False,\n",
       "  'snpCount': None,\n",
       "  'qualifier': None,\n",
       "  'imputed': True,\n",
       "  'pooled': False,\n",
       "  'studyDesignComment': None,\n",
       "  'accessionId': 'GCST90103972',\n",
       "  'fullPvalueSet': False,\n",
       "  'userRequested': False,\n",
       "  'platforms': [{'manufacturer': 'Illumina'}],\n",
       "  'ancestries': [{'type': 'initial',\n",
       "    'numberOfIndividuals': 15990,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'European'}],\n",
       "    'countryOfOrigin': [],\n",
       "    'countryOfRecruitment': [{'majorArea': 'NR',\n",
       "      'region': 'NR',\n",
       "      'countryName': 'NR'},\n",
       "     {'majorArea': 'Oceania',\n",
       "      'region': 'Australia/New Zealand',\n",
       "      'countryName': 'Australia'}]}],\n",
       "  'diseaseTrait': {'trait': 'Melanoma'},\n",
       "  'genotypingTechnologies': [{'genotypingTechnology': 'Genome-wide genotyping array'}],\n",
       "  'replicationSampleSize': 'NA',\n",
       "  'publicationInfo': {'pubmedId': '34813871',\n",
       "   'publicationDate': '2021-11-20',\n",
       "   'publication': 'J Invest Dermatol',\n",
       "   'title': 'Multi-trait genetic analysis identifies auto-immune loci associated with cutaneous melanoma.',\n",
       "   'author': {'fullname': 'Liyanage U', 'orcid': None}},\n",
       "  '_links': {'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST90103972'},\n",
       "   'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST90103972{?projection}',\n",
       "    'templated': True},\n",
       "   'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST90103972/associations?projection=associationByStudy'},\n",
       "   'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST90103972/snps'},\n",
       "   'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST90103972/backgroundEfoTraits'},\n",
       "   'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST90103972/efoTraits'},\n",
       "   'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST90103972/associations'}}},\n",
       " {'cohort': 'UKB',\n",
       "  'initialSampleSize': '3,122 White British ancestry cases, 271,708 White British ancestry controls',\n",
       "  'gxe': False,\n",
       "  'gxg': False,\n",
       "  'snpCount': 674885,\n",
       "  'qualifier': None,\n",
       "  'imputed': False,\n",
       "  'pooled': False,\n",
       "  'studyDesignComment': None,\n",
       "  'accessionId': 'GCST90027057',\n",
       "  'fullPvalueSet': False,\n",
       "  'userRequested': False,\n",
       "  'platforms': [],\n",
       "  'ancestries': [{'type': 'initial',\n",
       "    'numberOfIndividuals': 274830,\n",
       "    'ancestralGroups': [{'ancestralGroup': 'European'}],\n",
       "    'countryOfOrigin': [],\n",
       "    'countryOfRecruitment': [{'majorArea': 'Europe',\n",
       "      'region': 'Northern Europe',\n",
       "      'countryName': 'U.K.'}]}],\n",
       "  'diseaseTrait': {'trait': 'Melanoma'},\n",
       "  'genotypingTechnologies': [{'genotypingTechnology': 'Genome-wide genotyping array'}],\n",
       "  'replicationSampleSize': 'NA',\n",
       "  'publicationInfo': {'pubmedId': '34290314',\n",
       "   'publicationDate': '2021-07-21',\n",
       "   'publication': 'Sci Rep',\n",
       "   'title': 'Genetic association studies of alterations in protein function expose recessive effects on cancer predisposition.',\n",
       "   'author': {'fullname': 'Brandes N', 'orcid': None}},\n",
       "  '_links': {'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST90027057'},\n",
       "   'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST90027057{?projection}',\n",
       "    'templated': True},\n",
       "   'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST90027057/associations?projection=associationByStudy'},\n",
       "   'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST90027057/snps'},\n",
       "   'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST90027057/backgroundEfoTraits'},\n",
       "   'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST90027057/efoTraits'},\n",
       "   'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST90027057/associations'}}}]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import utils\n",
    "res = utils.analyze_json_keywords(metadata_list, [\"diabetes\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "12"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(metadata_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0.0, 100.0, {'diabetes': 12}, 0, 12, 12)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "______________________________________________________________________________________________________________\n",
      "Principle: F2\n",
      "Description: Findability: Keyword search in metadata\n",
      "Score: 0.8333333333333335\n",
      "Explanation: 83.33% of entities have all required fields (10 out of 12).\n",
      "16.67% of entities are missing some required fields (2 out of 12).\n",
      "\n",
      "Breakdown of missing fields:\n",
      "2 entities are missing 'melanoma' field.\n",
      "______________________________________________________________________________________________________________\n",
      "864 European ancestry cases, 864 European ancestry controls\n",
      "1,539 European ancestry cases, 3,917 European ancestry controls\n",
      "1,353 European ancestry cases, 3,566 European ancestry controls\n",
      "156 European ancestry cases from 34 high-risk families, 2,150 European ancestry controls\n",
      "2,168 European ancestry cases, 4,387 European ancestry controls\n",
      "2,804 European ancestry cases, 7,618 European ancestry controls\n",
      "1,804 European ancestry cases, 1,026 European ancestry controls\n",
      "494 European ancestry cases, 5,628 European ancestry controls\n",
      "4,842 European ancestry cases, 286,565 European ancestry controls\n",
      "6,777 European ancestry cases, 410,350 European ancestry controls\n",
      "4,089 European ancestry cases, 11,901 European ancestry controls\n",
      "3,122 White British ancestry cases, 271,708 White British ancestry controls\n",
      "______________________________________________________________________________________________________________\n",
      "Principle: F3\n",
      "Description: Findability: Check for identifiers\n",
      "Score: 1.0\n",
      "Explanation: 100.00% of entities have an ID field (12 out of 12).\n",
      "0.00% of entities are missing the ID field (0 out of 12).\n",
      "______________________________________________________________________________________________________________\n",
      "______________________________________________________________________________________________________________\n",
      "Principle: A1\n",
      "Description: Accessibility: Request status\n",
      "Score: 1\n",
      "Explanation: Successful request: API request returned the status 200.\n",
      "______________________________________________________________________________________________________________\n",
      "______________________________________________________________________________________________________________\n",
      "Principle: A1\n",
      "Description: Accessibility: Request status\n",
      "Score: 1\n",
      "Explanation: Successful request: API request returned the status 200.\n",
      "______________________________________________________________________________________________________________\n",
      "______________________________________________________________________________________________________________\n",
      "Principle: A1.1\n",
      "Description: Accessibility: Check after A1\n",
      "Score: 1\n",
      "Explanation: Follow-up check successful: API request returned the status 200.\n",
      "______________________________________________________________________________________________________________\n",
      "______________________________________________________________________________________________________________\n",
      "Principle: A1\n",
      "Description: Accessibility: Request status\n",
      "Score: 1\n",
      "Explanation: Successful request: API request returned the status 200.\n",
      "______________________________________________________________________________________________________________\n",
      "______________________________________________________________________________________________________________\n",
      "Principle: A1.1\n",
      "Description: Accessibility: Check after A1\n",
      "Score: 1\n",
      "Explanation: Follow-up check successful: API request returned the status 200.\n",
      "______________________________________________________________________________________________________________\n",
      "______________________________________________________________________________________________________________\n",
      "Principle: A1.2\n",
      "Description: Accessibility: Another check after A1\n",
      "Score: 1\n",
      "Explanation: Additional check successful: API request returned the status 200.\n",
      "______________________________________________________________________________________________________________\n",
      "(Meta)data is in format: json\n",
      "(Meta)data is in format: json\n",
      "(Meta)data is in format: json\n",
      "(Meta)data is in format: json\n",
      "(Meta)data is in format: json\n",
      "(Meta)data is in format: json\n",
      "(Meta)data is in format: json\n",
      "(Meta)data is in format: json\n",
      "(Meta)data is in format: json\n",
      "(Meta)data is in format: json\n",
      "(Meta)data is in format: json\n",
      "(Meta)data is in format: json\n",
      "______________________________________________________________________________________________________________\n",
      "Principle: I1\n",
      "Description: Interoperability: Format check\n",
      "Score: 1.0\n",
      "Explanation: Format is json.\n",
      "______________________________________________________________________________________________________________\n",
      "{'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000198'}, 'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000198{?projection}', 'templated': True}, 'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000198/associations?projection=associationByStudy'}, 'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000198/associations'}, 'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000198/snps'}, 'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000198/backgroundEfoTraits'}, 'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000198/efoTraits'}}\n",
      "{'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000437'}, 'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000437{?projection}', 'templated': True}, 'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000437/associations?projection=associationByStudy'}, 'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000437/snps'}, 'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000437/backgroundEfoTraits'}, 'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000437/efoTraits'}, 'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST000437/associations'}}\n",
      "{'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001886'}, 'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001886{?projection}', 'templated': True}, 'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001886/associations?projection=associationByStudy'}, 'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001886/associations'}, 'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001886/snps'}, 'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001886/backgroundEfoTraits'}, 'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001886/efoTraits'}}\n",
      "{'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001129'}, 'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001129{?projection}', 'templated': True}, 'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001129/associations?projection=associationByStudy'}, 'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001129/snps'}, 'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001129/backgroundEfoTraits'}, 'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001129/efoTraits'}, 'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001129/associations'}}\n",
      "{'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001266'}, 'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001266{?projection}', 'templated': True}, 'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001266/associations?projection=associationByStudy'}, 'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001266/associations'}, 'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001266/snps'}, 'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001266/backgroundEfoTraits'}, 'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001266/efoTraits'}}\n",
      "{'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001267'}, 'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001267{?projection}', 'templated': True}, 'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001267/associations?projection=associationByStudy'}, 'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001267/snps'}, 'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001267/backgroundEfoTraits'}, 'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001267/efoTraits'}, 'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001267/associations'}}\n",
      "{'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001245'}, 'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001245{?projection}', 'templated': True}, 'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001245/associations?projection=associationByStudy'}, 'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001245/associations'}, 'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001245/snps'}, 'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001245/backgroundEfoTraits'}, 'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001245/efoTraits'}}\n",
      "{'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST002514'}, 'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST002514{?projection}', 'templated': True}, 'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST002514/associations?projection=associationByStudy'}, 'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST002514/associations'}, 'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST002514/snps'}, 'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST002514/backgroundEfoTraits'}, 'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST002514/efoTraits'}}\n",
      "{'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST004142'}, 'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST004142{?projection}', 'templated': True}, 'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST004142/associations?projection=associationByStudy'}, 'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST004142/associations'}, 'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST004142/snps'}, 'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST004142/backgroundEfoTraits'}, 'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST004142/efoTraits'}}\n",
      "{'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST90011809'}, 'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST90011809{?projection}', 'templated': True}, 'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST90011809/associations?projection=associationByStudy'}, 'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST90011809/associations'}, 'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST90011809/snps'}, 'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST90011809/backgroundEfoTraits'}, 'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST90011809/efoTraits'}}\n",
      "{'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST90103972'}, 'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST90103972{?projection}', 'templated': True}, 'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST90103972/associations?projection=associationByStudy'}, 'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST90103972/snps'}, 'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST90103972/backgroundEfoTraits'}, 'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST90103972/efoTraits'}, 'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST90103972/associations'}}\n",
      "{'self': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST90027057'}, 'study': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST90027057{?projection}', 'templated': True}, 'associationsByStudySummary': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST90027057/associations?projection=associationByStudy'}, 'snps': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST90027057/snps'}, 'backgroundEfoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST90027057/backgroundEfoTraits'}, 'efoTraits': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST90027057/efoTraits'}, 'associations': {'href': 'https://www.ebi.ac.uk/gwas/rest/api/studies/GCST90027057/associations'}}\n",
      "______________________________________________________________________________________________________________\n",
      "Principle: I3\n",
      "Description: Interoperability: Metadata include qualified references to other (meta)data\n",
      "Score: 1.0\n",
      "Explanation: 100.00% of entities have a reference to other metadata.\n",
      "______________________________________________________________________________________________________________\n",
      "None\n",
      "None\n",
      "None\n",
      "None\n",
      "None\n",
      "None\n",
      "None\n",
      "None\n",
      "None\n",
      "None\n",
      "None\n",
      "None\n",
      "______________________________________________________________________________________________________________\n",
      "Principle: R1\n",
      "Description: Reusability: Metadata completeness\n",
      "Score: 0.0\n",
      "Explanation: Metadata is incomplete: it does not include the type of object nor the metadata specify the content of the data technical properties of its data file and format.\n",
      "______________________________________________________________________________________________________________\n",
      "______________________________________________________________________________________________________________\n",
      "Principle: R1.1\n",
      "Description: Reusability: Use of community standards\n",
      "Score: 1.0\n",
      "Explanation: All community standards fields present\n",
      "______________________________________________________________________________________________________________\n",
      "{'pubmedId': '18488026', 'publicationDate': '2008-05-18', 'publication': 'Nat Genet', 'title': 'Common sequence variants on 20q11.22 confer melanoma susceptibility.', 'author': {'fullname': 'Brown KM', 'orcid': None}}\n",
      "{'pubmedId': '19578364', 'publicationDate': '2009-07-05', 'publication': 'Nat Genet', 'title': 'Genome-wide association study identifies three loci associated with melanoma risk.', 'author': {'fullname': 'Bishop DT', 'orcid': None}}\n",
      "{'pubmedId': '23455637', 'publicationDate': '2013-03-03', 'publication': 'Nat Genet', 'title': 'A variant in FTO shows association with melanoma risk not due to BMI.', 'author': {'fullname': 'Iles MM', 'orcid': None}}\n",
      "{'pubmedId': '21706340', 'publicationDate': '2011-06-26', 'publication': 'Hum Genet', 'title': 'A unique genome-wide association analysis in extended Utah high-risk pedigrees identifies a novel melanoma risk variant on chromosome arm 10q.', 'author': {'fullname': 'Teerlink C', 'orcid': '0000-0002-1992-2326'}}\n",
      "{'pubmedId': '21983785', 'publicationDate': '2011-10-09', 'publication': 'Nat Genet', 'title': 'Genome-wide association study identifies a new melanoma susceptibility locus at 1q21.3.', 'author': {'fullname': 'Macgregor S', 'orcid': '0000-0001-6731-8142'}}\n",
      "{'pubmedId': '21983787', 'publicationDate': '2011-10-09', 'publication': 'Nat Genet', 'title': 'Genome-wide association study identifies three new melanoma susceptibility loci.', 'author': {'fullname': 'Barrett JH', 'orcid': None}}\n",
      "{'pubmedId': '21926416', 'publicationDate': '2011-09-22', 'publication': 'Hum Mol Genet', 'title': 'Genome-wide association study identifies novel loci predisposing to cutaneous melanoma.', 'author': {'fullname': 'Amos CI', 'orcid': None}}\n",
      "{'pubmedId': '24980573', 'publicationDate': '2014-06-30', 'publication': 'Carcinogenesis', 'title': 'Identification of a melanoma susceptibility locus and somatic mutation in TET2.', 'author': {'fullname': 'Song F', 'orcid': None}}\n",
      "{'pubmedId': '28212542', 'publicationDate': '2017-02-09', 'publication': 'Oncotarget', 'title': 'Two-stage genome-wide association study identifies a novel susceptibility locus associated with melanoma.', 'author': {'fullname': 'Ransohoff KJ', 'orcid': None}}\n",
      "{'pubmedId': '32887889', 'publicationDate': '2020-09-04', 'publication': 'Nat Commun', 'title': 'Pan-cancer study detects genetic risk variants and shared genetic basis in two large cohorts.', 'author': {'fullname': 'Rashkin SR', 'orcid': None}}\n",
      "{'pubmedId': '34813871', 'publicationDate': '2021-11-20', 'publication': 'J Invest Dermatol', 'title': 'Multi-trait genetic analysis identifies auto-immune loci associated with cutaneous melanoma.', 'author': {'fullname': 'Liyanage U', 'orcid': None}}\n",
      "{'pubmedId': '34290314', 'publicationDate': '2021-07-21', 'publication': 'Sci Rep', 'title': 'Genetic association studies of alterations in protein function expose recessive effects on cancer predisposition.', 'author': {'fullname': 'Brandes N', 'orcid': None}}\n",
      "______________________________________________________________________________________________________________\n",
      "Principle: R1.2\n",
      "Description: Reusability: Provenance metadata completeness\n",
      "Score: 1.0\n",
      "Explanation: All provenance metadata present\n",
      "______________________________________________________________________________________________________________\n",
      "╒════════════╤══════════╕\n",
      "│ Metric     │    Score │\n",
      "╞════════════╪══════════╡\n",
      "│ F2 score   │ 0.833333 │\n",
      "├────────────┼──────────┤\n",
      "│ F3 score   │ 1        │\n",
      "├────────────┼──────────┤\n",
      "│ A1 score   │ 1        │\n",
      "├────────────┼──────────┤\n",
      "│ A1.1 score │ 1        │\n",
      "├────────────┼──────────┤\n",
      "│ A1.2 score │ 1        │\n",
      "├────────────┼──────────┤\n",
      "│ I1 score   │ 1        │\n",
      "├────────────┼──────────┤\n",
      "│ I3 score   │ 1        │\n",
      "├────────────┼──────────┤\n",
      "│ R1 score   │ 0        │\n",
      "├────────────┼──────────┤\n",
      "│ R1.1 score │ 1        │\n",
      "├────────────┼──────────┤\n",
      "│ R1.2 score │ 1        │\n",
      "╘════════════╧══════════╛\n"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "'F1 score'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[8], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m scores \u001b[38;5;241m=\u001b[39m \u001b[43massess\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmetadata_list\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkeywords\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrepository_choice\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrequest_status\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/Desktop/ft/FAIR/assess.py:48\u001b[0m, in \u001b[0;36massess\u001b[0;34m(metadata, keywords, repository_choice, url, request_status)\u001b[0m\n\u001b[1;32m     44\u001b[0m     file\u001b[38;5;241m.\u001b[39mwrite(tabulate(data, headers\u001b[38;5;241m=\u001b[39m[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mMetric\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mScore\u001b[39m\u001b[38;5;124m\"\u001b[39m], tablefmt\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mplain\u001b[39m\u001b[38;5;124m\"\u001b[39m))\n\u001b[1;32m     46\u001b[0m scores \u001b[38;5;241m=\u001b[39m {row[\u001b[38;5;241m0\u001b[39m]: row[\u001b[38;5;241m1\u001b[39m] \u001b[38;5;28;01mfor\u001b[39;00m row \u001b[38;5;129;01min\u001b[39;00m data}\n\u001b[0;32m---> 48\u001b[0m \u001b[43mplot_fair_scores\u001b[49m\u001b[43m(\u001b[49m\u001b[43mscores\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     50\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m [row[\u001b[38;5;241m1\u001b[39m] \u001b[38;5;28;01mfor\u001b[39;00m row \u001b[38;5;129;01min\u001b[39;00m data]\n",
      "File \u001b[0;32m~/Desktop/ft/FAIR/assess.py:60\u001b[0m, in \u001b[0;36mplot_fair_scores\u001b[0;34m(scores)\u001b[0m\n\u001b[1;32m     53\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mplot_fair_scores\u001b[39m(scores):\n\u001b[1;32m     54\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m     55\u001b[0m \u001b[38;5;124;03m    Generate a bar graph for FAIR principles scores.\u001b[39;00m\n\u001b[1;32m     56\u001b[0m \n\u001b[1;32m     57\u001b[0m \u001b[38;5;124;03m    Args:\u001b[39;00m\n\u001b[1;32m     58\u001b[0m \u001b[38;5;124;03m        scores (dict): A dictionary of scores for each metric.\u001b[39;00m\n\u001b[1;32m     59\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m---> 60\u001b[0m     F_score \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39mmean([\u001b[43mscores\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mF1 score\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m, scores[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mF2 score\u001b[39m\u001b[38;5;124m'\u001b[39m], scores[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mF3 score\u001b[39m\u001b[38;5;124m'\u001b[39m], scores[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mF4 score\u001b[39m\u001b[38;5;124m'\u001b[39m]])\n\u001b[1;32m     61\u001b[0m     A_score \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39mmean([scores[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mA1 score\u001b[39m\u001b[38;5;124m'\u001b[39m], scores[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mA1.1 score\u001b[39m\u001b[38;5;124m'\u001b[39m], scores[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mA1.2 score\u001b[39m\u001b[38;5;124m'\u001b[39m], scores[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mA2 score\u001b[39m\u001b[38;5;124m'\u001b[39m]])\n\u001b[1;32m     62\u001b[0m     I_score \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39mmean([scores[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mI1 score\u001b[39m\u001b[38;5;124m'\u001b[39m], scores[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mI2 score\u001b[39m\u001b[38;5;124m'\u001b[39m], scores[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mI3 score\u001b[39m\u001b[38;5;124m'\u001b[39m]])\n",
      "\u001b[0;31mKeyError\u001b[0m: 'F1 score'"
     ]
    }
   ],
   "source": [
    "scores = assess(metadata_list, keywords, repository_choice, url, request_status)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
